---
title: "Phyloseq_RawData_Bacteria_PreProcess"
---

# Reference & Memo

\[R Package\] [Scales](https://scales.r-lib.org/)

[é£Ÿå“å¾®ç”Ÿç‰©å­¦ï¼ˆæ¤œæŸ»ã¨åˆ¶å¾¡æ–¹æ³•ï¼‰ï½œåŸºç¤ã¨æœ€æ–°æƒ…å ±ã‚’è§£èª¬ï½œæœ¨æ‘ã€€å‡¡](https://foodmicrob.com/16s-amplicon-sequencing-food-microbiology/)

[Phyloseq HP](https://joey711.github.io/phyloseq/)

<https://ushio-ecology-blog.blogspot.com/2021/04/20210428blogger0015.html>

[Bioconductor::MicrobiotaProcess](https://www.bioconductor.org/packages/release/bioc/html/MicrobiotaProcess.html)

[Introduction to MicrobiotaProcess](https://bioconductor.org/packages/devel/bioc/vignettes/MicrobiotaProcess/inst/doc//MicrobiotaProcess.html#biomarker-discovery)

[Github_MicrobiotaProcess](https://github.com/YuLab-SMU/MicrobiotaProcess)

[ggtreeExtra HP](https://bioconductor.org/packages/release/bioc/vignettes/ggtreeExtra/inst/doc/ggtreeExtra.html)

[Metagenomic biomarker discovery and explanation](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2011-12-6-r60)

::: callout-note
## DataLocation

ASVsTable (ä½•ã‚‚å‡¦ç†ã‚’ã—ã¦ã„ãªã„Table)

'\~/Library/CloudStorage/GoogleDrive-saito2022\@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/ASVsTable.csv'
:::

# NGS Analysis Flow

::: callout-note
## [Filtering]{.mark-cyan}

ä»¥ä¸‹ã€è«–æ–‡ã§ä½¿ã‚ã‚Œã¦ã„ã‚‹æ‰‹æ³•

ãƒ»ç›¸å¯¾å­˜åœ¨é‡ 1% \>\
ãƒ»å¹³å‡ç›¸å¯¾å­˜åœ¨é‡ 10\^-5 \>\
ãƒ»Prevalenceã§ã€sampleã®å†…ã€?ã§ã—ã‹ç¢ºèªã•ã‚Œã¦ã„ãªã„ASVs(OTUs)ã‚’æ’é™¤\
ãƒ»å‡¦ç†åŒº(è¾²è–¬æ•£å¸ƒ or ç—…å®³ç™ºç”Ÿã‚µãƒ³ãƒ—ãƒ«)ã§å­˜åœ¨ãŒç¢ºèªã•ã‚Œãªã„ASVsã¯æ’é™¤ãªã©

â†’ Filteringã®åŸºæº–ã¯ã€è«–æ–‡ã«ã‚ˆã£ã¦æ§˜ã€…ã§ã€ãŠãã‚‰ããã®æ™‚ã€…ã§éƒ½åˆãŒã„ã„ã‚ˆã†ã«è§£é‡ˆ??

Q: What is the difference between averaging, homogenization and standardization?

Averaging and homogenization both refer to the random sampling of sequences from different samples according to a certain rule (e.g., we go by the minimum number of sequences) so that the number of sequences in different samples is the same.

Standardization is generally a Z-value calculation for data that fluctuates over a large range, scaling the data so that it falls into a small specific interval that is easy to display in a graph.
:::

::: callout-note
## å¤šæ§˜æ€§è©•ä¾¡

â†’ æ•°åƒã®ASVs(OTU)ãŒæ¤œå‡ºã•ã‚Œã‚‹ãŸã‚ã€å„ASVsã«ç€ç›®ã™ã‚‹ã ã‘ã§ãªãã€ã‚µãƒ³ãƒ—ãƒ«å†…ã®å¾®ç”Ÿç‰©å¢(ASVså…¨ä½“)ã‚’æ§‹é€ ã‚„çµ„æˆã¨ã„ã£ãŸè¦³ç‚¹ã‹ã‚‰ã€è©•ä¾¡ã™ã‚‹ãŸã‚\
ã“ã‚Œã¾ã§å…ˆè¡Œç ”ç©¶ã§ã¯ã€æœ‰æ©Ÿè‚¥æ–™ãŒå¾®ç”Ÿç‰©æ•°ã‚„å¢ã‚’æ´»æ€§åŒ–ã•ã›ã¦ã€ç—…å®³ç™ºç”Ÿã‚’æŠ‘åˆ¶ã™ã‚‹ãªã©å ±å‘Šã•ã‚Œã¦ã„ã‚‹ãŒã€ã‚ã‚‹ç‰¹å®šã®å¾®ç”Ÿç‰©(ç´°èŒãƒ»ç³¸çŠ¶èŒ)ãŒå¯†æ¥ã«ç—…å®³ç™ºç”Ÿã«é–¢ã‚ã£ã¦ã„ã‚‹å¯èƒ½æ€§ã‚‚ã‚ã‚‹ãŒã€ãã‚Œã‚ˆã‚Šã‚‚è¤‡æ•°ã®å¾®ç”Ÿç‰©åŒå£«ãŒè¤‡é›‘ã«é–¢ã‚ã‚Šåˆã£ã¦ã€ç—…åŸèŒã®æ„ŸæŸ“ãŒã€ç«¶åˆçš„ã«åŠ£ã‚Šã€ç—…å®³ç™ºç”Ÿã«è‡³ã£ã¦ã„ãªã„å¯èƒ½æ€§ã‚‚ã‚ã‚‹ãŸã‚\
ã“ã®è¤‡é›‘æ€§ã‚’ç°¡å˜ã«çµ±è¨ˆçš„ã«ã‚‚ç§‘å­¦çš„ã«ã‚‚è©•ä¾¡ã™ã‚‹ã®ã¯ã€ä¸å¯èƒ½ã«è¿‘ã„ã“ã¨ã‚’å‰æã«ç½®ã„ãŸä¸Šã§ã€å¾®ç”Ÿç‰©å¢è§£æã®è«–æ–‡ãªã©ã‚’èª­ã‚€æ–¹ãŒè‰¯ã„ã€‚è§£æã¯ã€è‘—è€…ã®ä¸»è¦³ã‚‚å…¥ã‚‹ã“ã¨ã‚‚ã‚ã‚Šã€æ‰¹åˆ¤çš„ãªè¦–ç‚¹ã§èª­ã‚€ã“ã¨

ãƒ»Î²å¤šæ§˜æ€§(Tuley HSDæ¤œå®š)

ãƒ»The Shannon index curves & Rarefaction curves (b) of the bacterial community

Differences in fungal and prokaryotic community composition were tested through permutational multivariate analysis of variance (PERMANOVA) with the â€˜adonis2â€™ function on Bray-Curtis distances in the R package vegan \[45\]
:::

::: callout-note
## ç¾¤é›†æ§‹é€ ã®è©•ä¾¡ (community composition)

ãƒ»PCoA â†’ èª¿æŸ»ã—ãŸSampleæ•°(åœŸå£Œã‚µãƒ³ãƒ—ãƒ«(eDNA))ãŒå¤šã„æ™‚ã«ã¯ã€éå¸¸ã«æœ‰åŠ¹ã ãŒã€ã‚µãƒ³ãƒ—ãƒ«æ•°ãŒå°‘ãªã„å ´åˆã¯ã€è§£æã®å‰æãŒæˆç«‹ã—ãªã„ã“ã¨ã‚‚ã‚ã‚‹ã®ã§æ³¨æ„

ãƒ»NMDS(ANOSIM\[vegan\]ã§æœ‰æ„å·®æ¤œå®š

ãƒ»Differences in fungal and prokaryotic community composition were tested through permutational multivariate analysis of variance (PERMANOVA) with the â€˜adonis2â€™ function on Bray-Curtis distances in the R package vegan
:::

::: callout-note
## Differentially abundant taxa

[**æœ‰æ„ã«å­˜åœ¨é‡ãŒç•°ãªã‚‹åˆ†é¡ç¾¤(ASVs)ã®æ¢ç´¢**]{.mark-blue}

ãƒ»ç›¸å¯¾å­˜åœ¨é‡ã®æ¯”è¼ƒ(one-way ANOVA, LSD)\
â†’ ä¸»è¦ãªåˆ†é¡ç¾¤ã®ç‰¹å¾´ã‚’æ¢ç´¢

ãƒ»ãƒ©ãƒ³ãƒ€ãƒ ãƒ•ã‚©ãƒ¬ã‚¹ãƒˆãƒ¢ãƒ‡ãƒ«ã§ã€è¾²è–¬æ•£å¸ƒã«ã‚ˆã£ã¦å¤‰å‹•ã—ãŸå¾®ç”Ÿç‰©å¢ã«åå¿œã—ãŸåˆ†é¡ç¾¤(ASVs)ã‚’äºˆæ¸¬\
We built random forest regression models to test the effect of altered prokaryote abundance through fungal diversity by using prokaryote abundances to predict fungal diversity.

Random forest models were generated with theâ€˜randomForestâ€™ function in the randomForest R package \[49\]. To remove redundant features and avoid over tting models, we removed redundant OTUs with theâ€˜Borutaâ€™ function in the package Boruta \[50\].

The method performs a top-down search for relevant OTUs by comparing the importance of the original OTUs from those selected at random. Models were tuned to achieve the lowest stable out-of-bag (OOB) error estimate possible, and the best mtry value (number of OTUs sampled at random in the entire pool for each tree at each split) was selected using theâ€˜tuneRFâ€™ function in randomForest R package.

ãƒ»Fungal OTUs were determined to be signicant if the W value was greater than 70% of the taxa tested based on Wilcoxon ranked sum test between additive log-ratio transformed data and a Benjamini- Hochbergj adjusted P value (Î± = 0.05) \[47\].

ãƒ»ANCOMï¼ˆAnalysis of Composition of Microbiomesï¼‰

âœ… å®šç¾© ANCOMã¨ã¯ã€å¾®ç”Ÿç‰©ç¾¤é›†ã«ãŠã‘ã‚‹ç¾¤é–“ã®å·®ç•°ï¼ˆä¾‹ãˆã°å‡¦ç†åŒºã¨å¯¾ç…§åŒºãªã©ï¼‰ã‚’ã€ç›¸å¯¾çš„ãªãƒ‡ãƒ¼ã‚¿ï¼ˆcompositional dataï¼‰ã«åŸºã¥ã„ã¦è§£æã™ã‚‹æ–¹æ³•ã§ã™ã€‚ ç‰¹ã«ã€16S rRNAã‚·ãƒ¼ã‚±ãƒ³ã‚¹ãªã©ã®ãƒ‡ãƒ¼ã‚¿ã§ã¯ã€ç·é‡ï¼ˆlibrary sizeï¼‰ã«åˆ¶é™ãŒã‚ã‚‹ãŸã‚ã€å¾—ã‚‰ã‚Œã‚‹ãƒ‡ãƒ¼ã‚¿ã¯ã€Œç›¸å¯¾çš„ã€ãªã‚‚ã®ã§ã‚ã‚‹ã“ã¨ã‹ã‚‰ã€é€šå¸¸ã®çµ±è¨ˆè§£æã§ã¯èª¤å·®ã‚„ãƒã‚¤ã‚¢ã‚¹ãŒç”Ÿã˜ã‚„ã™ããªã‚Šã¾ã™ã€‚\
ANCOMã¯ãã®ç‚¹ã‚’è€ƒæ…®ã—ã€ç›¸å¯¾ãƒ‡ãƒ¼ã‚¿ã«é©ã—ãŸæ–¹æ³•ã§ã€Œã©ã®åˆ†é¡ç¾¤ï¼ˆèŒç¨®ã‚„å±ãªã©ï¼‰ãŒæœ‰æ„ã«ç•°ãªã‚‹ã‹ã€ã‚’æ¤œå‡ºã™ã‚‹æ‰‹æ³•ã¨ã—ã¦é–‹ç™ºã•ã‚Œã¾ã—ãŸã€‚

BioManeger::ANCOMBS package

ãƒ»LEfSeè§£æ â†’ æ¯”è¼ƒã—ãŸã„å‡¦ç†åŒºã§ã€æœ€ä½ã§ã‚‚5ã¤ä»¥ä¸Šã¯ã‚µãƒ³ãƒ—ãƒ«ãŒå¿…è¦ãªã®ã§ã€æ³¨æ„!!!
:::

# Paper Information

::: callout-note
## Waste not, want not: revisiting the analysis that called into question the practice of rarefaction

they removed rare and low-prevalence OTUs in two steps.

First, they removed any OTUs whose total abundance was less than three across all 80 samples and that did not appear in at least three samples.

Second, they removed any OTUs that did not have more than one sequence in more than 5% of the 80 samples (i.e., four samples) and that did not have a total abundance across the 80 samples greater than one half of the number of samples in each community type (i.e., 20)

Relative abundance data were used to calculate Bray-Curtis, Unweighted UniFrac, and Weighted UniFrac distances

ã€Discussionã€‘

It is worth commenting on WNWNâ€™s advice to use DESeqâ€™s Variance Stabilization or edgeRâ€™s Upper-Quartile Log-fold Change normalization strategies. These methods have been adopted from gene expression analysis to microbiome analysis.

Gene expression analysis implicitly assumes that all samples have the same genes. While this assumption might be valid when comparing host gene expression in healthy and diseased tissues from a cohort of patients, it does not generalize to those patientsâ€™ microbiota.

Microbial populations are highly patchy in their distribution. Thus, a zero count for gene expression is more likely to represent a gene below the limit of detection, whereas a zero count for a microbiome analysis is more likely to represent the true absence of the OTU.

In fact, zeroes are less common in gene expression analyses. However, because microbiome studies have so many zeroes, it is necessary to add a pseudo-count to all OTUs for both the edgeR- and DESeq-based normalizations (L443 and L487, respectively).

In WNWN, a pseudo-count of 1 was used. However, this value is arbitrary and the results can vary based on the choice of pseudo-count and the patchiness of the communities being analyzed.

Since WNWN was published, compositional approaches have been proposed to account for uneven sequencing and to provide improved interpretability (14, 16â€“18, 44, 45). However, these methods also often require the use of pseudo-counts. Rarefaction is preferred to these alternatives.

[The choice of distance metric is a complicated question]{.mark-green} â†’ Note_KeyNotes (SCISPACE)ã‚’å‚ç…§ â†’ RarefactionãŒã€ã‚µãƒ³ãƒ—ãƒ«ã§ç•°ãªã‚‹ã‚·ãƒ¼ã‚±ãƒ³ã‚¹æ·±åº¦ã®å½±éŸ¿ã‚’åˆ¶å¾¡ã™ã‚‹å”¯ä¸€ã®æ–¹æ³•ã§ã‚ã‚‹ã‹ã‚‚ã—ã‚Œãªã„...ã€ç›¸å¯¾å­˜åœ¨é‡ã®åˆ©ç”¨ã«ã‚‚æ³¨æ„ãŒå¿…è¦ â†’ Referenceã®è«–æ–‡ã‚‚èª­ã‚“ã§ã¿ã‚‹
:::

::: callout-note
## [Non-target impacts of fungicide disturbance on phyllosphere yeasts in conventional and no-till management](https://www.nature.com/articles/s43705-022-00103-w#Sec15)

**\[Suppliments\]**

Import and Preprocessing in R. Data were imported into R 4.0.3 \[17\], and the R packages phyloseq 1.24.2 \[18\] and vegan 2.5.3 \[19\] were used for most analyses.

**Samples with low sequencing coverage (less than 1000 reads) were removed from the analysis.**

Contaminant OTUs (i.e., those prevalent in negative extraction controls) were removed with the R package decontam \[20\].

Before normalization, richness was assessed for Prokaryotes and Fungi in the leaves and roots of each crop using the â€˜estimate_richnessâ€™ function in the phyloseq package.

Results of alpha diversity analyses were plotted using the â€˜ggplot2â€™ package \[21\].

Then, sample read counts were normalized using the cumulative sum scaling technique within the metagenomeSeq R package \[22\].

**\[Statistical analysis:\]**

For differential abundance analysis, fungal OTUs (fOTU, hereafter) with a [**mean relative abundance less than 10\^âˆ’5**]{.underline} and [**fOTUs with zeros present in 95% samples**]{.underline} were discarded from the analysis to avoid detecting fOTUs as signi cantly different based on stochasticity.\
[**In addition, fOTUs that were never present in fungicide treated plots were not included.**]{.underline}\
â†’ Prevalence + ç›¸å¯¾å­˜åœ¨é‡ã§Filtering + è¾²è–¬å‡¦ç†åŒºã§å­˜åœ¨ã—ã¦ã„ãªã„OTUsã¯é™¤å¤–ã—ã¦ã€å®Ÿæ–½
:::

::: callout-note
## MicrobiotaProcess: A comprehensive R package for deep mining microbiome
:::

::: callout-note
## Non-target impacts of fungicide disturbance on phyllosphere yeasts in conventional and no-till management

\[Statistical analysis:\]

Differences in fungal and prokaryotic community composition were tested through permutational multivariate analysis of variance (PERMANOVA) with the â€˜adonis2â€™ function on Bray-Curtis distances in the R package vegan

Differentially abundant taxa resulting from fungicide application were identi ed by comparing fungicide treated plots to control plots through an analysis of the composition of microbiomes (ANCOM v 2.1)\
**â†’ å‡¦ç†åŒºã«å¿…è¦ãªã‚µãƒ³ãƒ—ãƒ«æ•°ãŒ5ã¤ä»¥ä¸Š**\
ã€Analysis of composition of microbiomes: a novel method for studying microbial compositionã€‘
:::

# Chunk Options & SetDirectory

```{r}
#| label: ChunkOption
#| include: true
#| message: false

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome")

library("MicrobiotaProcess")
library("phyloseq")
library("dplyr")
library("purrr")
library("ggplot2")
library("scales")

# https://astrobiomike.github.io/amplicon/dada2_workflow_ex

```

# MergeingMethods:

::: callout-note
MergingMethods

Merging methods include `merge_taxa` and `merge_samples`, intended for merging specific OTUs or samples, respectively.

There is also the `merge_phyloseq` function for a complete merge of two or more phyloseq-objects (or a phyloseq-object and one or more separate components). For example, the following code merges the first 5 OTUs in the Chlamydiae-only dataset.

**åˆ†é¡ãƒ¬ãƒ™ãƒ«ã®çµ±ä¸€**

OTU(or ASVs)ã¯é€šå¸¸ã€**ã§ãã‚‹ã ã‘ç´°ã‹ã„åˆ†é¡ï¼ˆç¨®ãƒ¬ãƒ™ãƒ«ãªã©ï¼‰** ã§åˆ†ã‘ã‚‰ã‚Œã‚‹ãŒã€å ´åˆã«ã‚ˆã£ã¦ã€**ä¸Šä½åˆ†é¡ï¼ˆå±ãƒ¬ãƒ™ãƒ«ã‚„ç§‘ãƒ¬ãƒ™ãƒ«ï¼‰ã§è§£æã‚’è¡Œã†æ–¹ãŒé©ã—ã¦ã„ã‚‹** ã“ã¨ãŒã‚ã‚‹ã€‚

> å±ã‚„ç§‘ã§ã€RelativeAbundunceã‚’plotã™ã‚‹å ´åˆãªã©ã«åˆ©ç”¨ã™ã‚‹

âœ… **ç‰¹å®šã®åˆ†é¡ãƒ¬ãƒ™ãƒ«ï¼ˆä¾‹: å±ãƒ¬ãƒ™ãƒ«ï¼‰ã§ã®è§£æã‚’è¡Œã†ãŸã‚ã« OTU ã‚’çµ±åˆã™ã‚‹**

-   ä¾‹: **åŒã˜å± (`Genus`) ã«å±ã™ã‚‹ OTU ã‚’ã¾ã¨ã‚ã‚‹**

-   ä¾‹: **é–€ (`Phylum`) ã‚„ã‚¯ãƒ©ã‚¹ (`Class`) ã§ãƒ‡ãƒ¼ã‚¿ã‚’æ•´ç†ã™ã‚‹**

> **ğŸ”¹ ä¾‹ãˆã°**
>
> -   `OTU_1` â†’ *Bacteroides fragilis*
>
> -   `OTU_2` â†’ *Bacteroides vulgatus*
>
>     â†’ **ã“ã‚Œã‚‰ã‚’ "Bacteroides å±" ã¨ã—ã¦çµ±åˆã™ã‚‹ã€ã¤ã¾ã‚Šä¸Šä½ã®åˆ†é¡ãƒ¬ãƒ™ãƒ«ã§æ¯”è¼ƒã—ãŸã„å ´åˆã«ã€ä½¿ç”¨ã™ã‚‹ã€ç‰¹å®šã®ASVsã‚’çµ±åˆã™ã‚‹å ´åˆã«ä½¿ç”¨ã™ã‚‹**
>
>     â†’ å…¨ã¦ã®å±ã‚’çµ±åˆã™ã‚‹ãªã‚‰ã€ä»¥ä¸‹ã®ã‚ˆã†ã«åˆ†é¡å˜ä½ã‚’æŒ‡å®šã™ã‚Œã°è‰¯ã„\
>     `tax_glom(physeqã‚ªãƒ–ã‚¸ã‚§ã‚¯ãƒˆ, taxrank = "Family", NArm = true/false)`
>
>     â†’ [**çµ±åˆå¾Œã¯ã€çµ±åˆã—ãŸåˆ†é¡ãƒ¬ãƒ™ãƒ«ä»¥é™ã¯ã€NAã§è¡¨ç¤ºã•ã‚Œã‚‹**]{.mark-blue}

------------------------------------------------------------------------

`data("GlobalPatterns")`

`GP.chl = subset_taxa(GlobalPatterns, Phylum=="Chlamydiae") GP.chl = prune_samples(sample_sums(GP.chl)>=20, GP.chl)`

`GP.chl.merged = merge_taxa(GP.chl, taxa_names(GP.chl)[1:5])`

`ExportCSVData_GP.chl <- cbind(GP.chl@otu_table@.Data, GP.chl@tax_table@.Data)`

`write.csv(ExportCSVData_GP.chl, file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/ExportCSVData_GP.chl.csv", row.names = TRUE)`

`ExportCSVData_GP.chl.merged <- cbind(GP.chl.merged@otu_table@.Data, GP.chl.merged@tax_table@.Data)`

`write.csv(ExportCSVData_GP.chl.merged, file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/ExportCSVData_GP.chl.merged.csv", row.names = TRUE)`

çµ±åˆå‰

![GP.chl](Phyloseq_RawData_Bac_images/tax_glom_v1.png)

çµ±åˆå¾Œ

24341(ASVs)ã«çµ±åˆã•ã‚Œã€ãƒªãƒ¼ãƒ‰æ•°ãŒåˆè¨ˆã•ã‚Œã‚‹

![GP.chl.merge](Phyloseq_RawData_Bac_images/tax_glom_v2.png)
:::

# PreProcess Methods

## PreProcess Methods Scripts

## Prevalence filtering:

```{r}
#| label: Prevalence filtering
#| eval: false
#| Include: false


# Sample(Bac1~9)ã«ãŠã„ã¦ã€å„ASVsãŒå‡ºç¾ã—ãŸSampleæ•°(å­˜åœ¨é »åº¦)ã‚’å‡ºåŠ›
# taxa_are_rows â†’ ASVsãŒè¡ŒãŒåˆ—ã‹ã©ã†ã‹ã‚’ç¢ºèª
prev0 = apply(X = otu_table(PhyseqData),
              MARGIN = ifelse(taxa_are_rows(PhyseqData), yes = 1, no = 2),
              FUN = function(x){sum(x > 0)})

# DataFrameã®ä½œæˆ
prevdf = data.frame(Prevalence = prev0,
                    TotalAbundance = taxa_sums(PhyseqData),
                    tax_table(PhyseqData))

# å…¨ASVsã®ãƒ™ã‚¯ãƒˆãƒ«(ID)ã‚’å–ã‚Šå‡ºã™ â†’ TotalReads(ASVs) > 10ã§ãƒ•ã‚£ãƒ«ã‚¿ãƒªãƒ³ã‚°
# keep_taxa <- taxa_names(PhyseqData)[taxa_sums(PhyseqData) >= 10]
# ps_filt <- prune_taxa(keep_taxa, PhyseqData)


# é–€(Phylum)ãƒ¬ãƒ™ãƒ«ã§ã€tableã‚’å‡ºåŠ›ã€å„Phylumãƒ¬ãƒ™ãƒ«ã§åˆ†é¡ã•ã‚ŒãŸãƒ™ã‚¯ãƒˆãƒ«ã«å¯¾ã—ã¦ã€[ASVsæ•°>5]ã§Filtering
# ã¤ã¾ã‚Šã€ASVsãŒå…¨Sampleã®ä¸­ã§ã€5å›ä»¥ä¸Šå‡ºç¾ã—ã¦ã„ãªã„ã¨ã„ã‘ãªã„[ASVs1~5 = "Pseudomas"]çš„ãª
keepPhyla = table(prevdf$Phylum)[(table(prevdf$Phylum) > 5)]

prevdf1 = subset(prevdf, Phylum %in% names(keepPhyla))

# å…¨sampleæ•°ã®ã†ã¡ã€ä½•sampleã§ASVsãŒå‡ºç¾ã—ã¦ã„ã‚Œã°ã€è‰¯ã„ã‹ã®åŸºæº–ã‚’è¨­å®š
# 0.1ã®å‰²åˆã ã¨ã€9sampleã®å ´åˆã€1sampleã§å‡ºç¾ã—ã¦ã„ã‚Œã°ã„ã„ã“ã¨ã«ãªã‚‹ãŸã‚ã€1ã§è¨­å®š
prevalenceThreshold = 1
prevalenceThreshold


# Execute prevalence filter, using `prune_taxa()` function
ps1 = prune_taxa((prev0 > prevalenceThreshold), PhyseqData)
ps1


ps2 = subset_taxa(ps1, Phylum %in% names(keepPhyla))
ps2

Taxa_prevalence_total_counts <- 
    ggplot(prevdf1, aes(TotalAbundance, Prevalence, color = Phylum)) +
    geom_hline(yintercept = prevalenceThreshold, alpha = 0.5, linetype = 2) +
    geom_point(size = 2, alpha = 0.7) +
    scale_y_log10() + scale_x_log10() +
    xlab("Total Abundance") +
    facet_wrap(~Phylum) +
    theme(legend.position = "none")

Taxa_prevalence_total_counts

# save(Taxa_prevalence_total_counts,
#      file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/250331/RData/PlotPhyloseqData/Taxa_prevalence_total_counts.RData")


## Agglomerate closely related taxa ----------


# rank_names(PhyseqData)
# 
# # ps2ã¯ã€ãƒ•ã‚£ãƒ«ã‚¿ãƒªãƒ³ã‚°ã•ã‚Œã¦ããŸphyseqObjects
# ID <- rank_names(PhyseqData)
# for (i in ID) {
#     print(length(get_taxa_unique(ps2, taxonomic.rank = i)))
# }
# 
# ps3 = tax_glom(ps2, taxrank = "Genus")


## Abundance value transformation ------------

# meltphyseq = psmelt(ps3)
# meltphyseq_v1 <- meltphyseq |> filter(Phylum == "Pseudomonadota")

# write.csv(meltphyseq_v1, 
#           file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/meltphyseq.csv")

# orders <- unique(meltphyseq$Order)
# plot_list <- list()
# 
# for (order in orders) {
#     filt <- meltphyseq |> filter(Order == order)
#     
#     p <- ggplot(data = filt, mapping = aes(x = dps, y = Abundance)) +
#         # ãƒã‚¤ã‚ªãƒªãƒ³ãƒ—ãƒ­ãƒƒãƒˆã®è¨­å®š
#         geom_violin(aes(fill = dps), alpha = 0.3, position = position_nudge(x = 0.1)) +  # ãƒã‚¤ã‚ªãƒªãƒ³ã‚’å°‘ã—ãšã‚‰ã™
#         # ãƒã‚¤ãƒ³ãƒˆã®è¨­å®š
#         geom_point(aes(color = Sample.Name), size = 3, alpha = 0.8) + 
#         # ãƒã‚¤ãƒ³ãƒˆã®è‰²ã‚’Sample.Nameã«åŸºã¥ã„ã¦å¤‰æ›´
#         ylab(paste(order)) +  # yè»¸ãƒ©ãƒ™ãƒ«ã«Orderåã‚’ä½¿ç”¨
#         scale_y_log10() +  # yè»¸ã‚’ãƒ­ã‚°ã‚¹ã‚±ãƒ¼ãƒ«
#         scale_x_continuous(breaks = c(0, 3, 7)) +  # xè»¸ã®ãƒ©ãƒ™ãƒ«è¨­å®š
#         theme_minimal()  # ã‚°ãƒ©ãƒ•ã®ãƒ†ãƒ¼ãƒã‚’è¨­å®š
#     
#     # ãƒ—ãƒ­ãƒƒãƒˆã‚’ãƒªã‚¹ãƒˆã«æ ¼ç´
#     plot_list[[order]] <- p
# }


## æœ€åˆã®ãƒ—ãƒ­ãƒƒãƒˆã‚’è¡¨ç¤º
# print(plot_list[[1]])
# 
# length(get_taxa_unique(ps2, taxonomic.rank = "Phylum"))


# prune_samples(sample_sums(physeqData_???)>=20, physeqData_???)
# ExportCSVData <- cbind(t(PhyseqData@otu_table@.Data), PhyseqData@tax_table@.Data)
# write.csv(ExportCSVData, file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/ExportCSVData.csv",
#           row.names = TRUE)

# Remove taxa not seen more than 3 times in at least 20% of the samples. This protects against an OTU with small mean & trivially large C.V.
# filter_taxa(PhyseqData, function(x) sum(x > 3) > (0.2*length(x)), TRUE)


# Standardize abundances to the median sequencing depth
# sampleé–“ã§ç•°ãªã‚‹ãƒªãƒ¼ãƒ‰æ•°(ã‚·ãƒ¼ã‚±ãƒ³ã‚¹æ·±åº¦)ã‚’ä¸€è‡´ã•ã›ã‚‹ãŸã‚
# total = median(sample_sums(physeqData))
# standf = function(x, t=total) round(t * (x / sum(x)))
# physeqData_sd = transform_sample_counts(physeqData, standf)


# Filter the taxa using a cutoff of 3.0 for the Coefficient of Variation
# physeqData_cv = filter_taxa(PhyseqData_???, function(x) sd(x)/mean(x) > 3.0, TRUE)


```

## Filtering Methods:

```{r}
#| label: PreProcess & Plot & save
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center


load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/SaveObjects/PhyloseqData.RData")

PhyseqData
MetaData
track

colors <- c(
    "#1b9e77",  # ç·‘ç³»ï¼ˆæ¿ƒï¼‰
    "#d95f02",  # ã‚ªãƒ¬ãƒ³ã‚¸ç³»
    "#7570b3",  # é’ç´«ç³»
    "#e7298a",  # ãƒ”ãƒ³ã‚¯ç³»ï¼ˆæ˜ï¼‰
    "#66a61e",  # é»„ç·‘ç³»
    "#e6ab02",  # é»„åœŸè‰²
    "#a6761d",  # èŒ¶è‰²
    "#666666",  # ã‚°ãƒ¬ãƒ¼ï¼ˆæ¿ƒï¼‰
    
    "#8dd3c7",  # è–„ã„é’ç·‘
    "#ffffb3",  # è–„ã„é»„è‰²
    "#bebada",  # è–„ã„ç´«
    "#fb8072",  # è–„ã„èµ¤
    "#80b1d3",  # é’
    "#fdb462",  # ã‚ªãƒ¬ãƒ³ã‚¸
    "#b3de69",  # æ˜ã‚‹ã„ç·‘
    "#fccde5",  # è–„ã„ãƒ”ãƒ³ã‚¯
    
    "#bc80bd",  # ç´«
    "#ccebc5",  # ãƒŸãƒ³ãƒˆã‚°ãƒªãƒ¼ãƒ³
    "#ffed6f",  # ãƒ¬ãƒ¢ãƒ³ã‚¤ã‚¨ãƒ­ãƒ¼
    "#7fc97f",  # è½ã¡ç€ã„ãŸç·‘
    "#fdc086",  # æŸ”ã‚‰ã‹ã„ã‚ªãƒ¬ãƒ³ã‚¸
    "#ffff99",  # æ˜ã‚‹ã„é»„
    "#386cb0",  # ç´ºé’
    "#f0027f",  # ãƒ“ãƒ“ãƒƒãƒ‰ãƒ”ãƒ³ã‚¯
    
    "#bf5b17",  # èµ¤èŒ¶
    "#6a3d9a",  # æ¿ƒã„ç´«
    "#cab2d6",  # æ·¡ã„ãƒ©ãƒ™ãƒ³ãƒ€ãƒ¼
    "#ff7f00",  # é®®ã‚„ã‹ãªã‚ªãƒ¬ãƒ³ã‚¸
    "#b2df8a",  # æ˜ã‚‹ã„ç·‘
    "#a6cee3",  # æ˜ã‚‹ã„æ°´è‰²
    "#fb9a99",  # ã‚½ãƒ•ãƒˆãªèµ¤
    "#1f78b4",  # è½ã¡ç€ã„ãŸé’
    "#33a02c",  # æ·±ç·‘
    "#b15928"   # ç„¦ã’èŒ¶
)


color_v2 <- c("#1F77B4", "#FF7F0E", "#2CA02C", "#D62728", "#9467BD", "#8C564B",
              "#E377C2", "#7F7F7F", "#BCBD22", "#17BECF", "#F0E442", "#9E14D2",
              "#7D8B33", "#B5F56D", "#3F51B5", "#D32F2F", "#0288D1", "#7B1FA2", 
              "#388E3C", "#FBC02D", "#0288D1", "#8BC34A", "#FF5722", "#8E24AA", 
              "#795548", "#9C27B0", "#3F51B5", "#4CAF50", "#FF9800", "#E91E63", 
              "#CDDC39"
)


# FilteringMethods --------------------------

# Family Level 

PhyseqData_Family <- PhyseqData  |> 
    tax_glom(taxrank = "Family") %>%                        # agglomerate at phylum level
    transform_sample_counts(function(x) {x/sum(x)} )  |>    # Transform to relative abundance
    psmelt()  |>                                            # Melt to long format
    filter(Abundance > 0.01)  |>                            # Filter out low(>1%) abundance taxa
    arrange(desc(Family))                                   # Sort data frame alphabetically by phylum

unique(PhyseqData_Family$Family) 

# write.csv(PhyseqData_Family,
#           file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/PhyseqData_Family.csv", 
#           row.names = TRUE) 


# Genus Level 

PhyseqData_Genus <- PhyseqData  |> 
    tax_glom(taxrank = "Genus") |>                        
    transform_sample_counts(function(x) {x/sum(x)} )  |>   
    psmelt()  |>                                           
    filter(Abundance > 0.01)  |>                           
    arrange(desc(Genus))

PhyseqData_Genus$dps <- factor(PhyseqData_Genus$dps, levels = c(0, 3, 7))


# Objectã¨ã—ã¦ç’°å¢ƒã«æ ¼ç´ã—ã¦ã„ãªãã¦ã‚‚ã€x=psmelt(PhyseqData)ã§ã€æŒ‡å®šå¯èƒ½
## Tips: âŒ˜ + â‡§ + C â†’ ã‚³ãƒ¡ãƒ³ãƒˆã‚¢ã‚¦ãƒˆè§£é™¤

# write.csv(x = PhyseqData_Genus,
#           file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/PhyseqData_Genus.csv", 
#           row.names = TRUE) 

unique(PhyseqData_Genus$Genus) 



# Plot_Relative_Abundunce -------------------

# Family Level 

## Gathering dps
ggplot(PhyseqData_Family, aes(x = dps, y = Abundance, fill = Family)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors) +
    theme(axis.title.x = element_blank()) + # Remove x axis title
    guides(fill = guide_legend(reverse = F, keywidth = 1, keyheight = 1)) +
    scale_y_continuous(labels = percent) + # Package"scales"ãŒå¿…è¦
    xlab("days post Fungicide Inoculum") +
    ylab("Relative Abundance (Family > 1%) \n") +
    ggtitle("Relative abundance")

# ggsave(filename = "Relative_abundance_Family.png", plot = last_plot(),
#        width = 2000, height = 1800, dpi = 300, units = "px",
#        path = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/Figure")

## In Sample.Name
ggplot(PhyseqData_Family, aes(x = Sample.Name, y = Abundance, fill = Family)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors) +
    theme(axis.title.x = element_blank()) + # Remove x axis title
    guides(fill = guide_legend(reverse = F, keywidth = 1, keyheight = 1)) +
    scale_y_continuous(labels = percent) +
    xlab("days post Fungicide Inoculum") +
    ylab("Relative Abundance (Family > 1%) \n") +
    ggtitle("Relative abundance")

# ggsave(filename = "Relative_abundance_Family_SampleName.png", plot = last_plot(),
#        width = 2000, height = 1800, dpi = 300, units = "px",
#        path = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/Figure")


# Genus Level 

## Gathering dps
ggplot(PhyseqData_Genus, aes(x = dps, y = Abundance, fill = Genus)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors) +
    theme(axis.title.x = element_blank()) + # Remove x axis title
    guides(fill = guide_legend(reverse = F, keywidth = 1, keyheight = 1)) +
    scale_y_continuous(labels = percent) +
    xlab("days post Fungicide Inoculum") +
    ylab("Relative Abundance (Genus > 1%) \n") +
    ggtitle("Relative abundance")

# ggsave(filename = "Relative_abundance_Genus.png", plot = last_plot(),
#        width = 2000, height = 1800, dpi = 300, units = "px",
#        path = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/Figure")

## In Sample.Name
ggplot(PhyseqData_Genus, aes(x = Sample.Name, y = Abundance, fill = Genus)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors) +
    theme(axis.title.x = element_blank()) + # Remove x axis title
    guides(fill = guide_legend(reverse = F, keywidth = 1, keyheight = 1)) +
    scale_y_continuous(labels = percent) +
    xlab("days post Fungicide Inoculum") +
    ylab("Relative Abundance (Genus > 1%) \n") +
    ggtitle("Relative abundance")

# ggsave(filename = "Relative_abundance_Genus_SampleName.png", plot = last_plot(),
#        width = 2000, height = 1800, dpi = 300, units = "px",
#        path = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/Figure")

rm(list = ls())
```

::: callout-warning
## Abundance value transformation

[**It is usually necessary to transform microbiome count data to account for differences in library size, variance, scale, etc.**]{.mark-green}

`transform_sample_counts`:

The first argument to this function is the **phyloseq object** you want to transform, and the second argument is an R function that defines the transformation.

The R function is applied **sample-wise**, expecting that the first unnamed argument is a vector of taxa counts in the same order as the phyloseq object.

Additional arguments are passed on to the function specified in the second argument, providing an explicit means to include precomputed values, previously defined parameters/thresholds, or any other object that might be appropriate for computing the transformed values of interest.
:::

# MicrobiotaProcessMethods:

```{r}
#| label: MicrobiotaProcessMethods
#| eval: true
#| Include: false
#| echo: true

# MPSE object Constructs
# mpdata <- PhyseqData |>
#     as.MPSE()


library("MicrobiotaProcess")
library("phyloseq")
library("dplyr")
library("purrr")
library("ggplot2")

load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/SaveObjects/mpdata.RData")

mpdata %<>% mp_rrarefy()

mpdata %<>% 
    mp_cal_rarecurve(
        .abundance = RareAbundance,
        chunks = 400
    )

```

## alpha diversity analysis

[**ãƒ©ãƒ³ãƒ€ãƒ ã‚µãƒ³ãƒ—ãƒªãƒ³ã‚°ã•ã‚ŒãŸã®ãŒã€RareAbundance**]{.mark-red}

```{r}
#| label: alpha diversity analysis
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center



# mpdata <- PhyseqData |>
#     as.MPSE()

# save(mpdata,
#      file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/250331/RData/SaveObjects/mpdata.RData")


# ä»¥ä¸‹ã¯åŒã˜æ„å‘³
## x <- x |> log()
## x %<>% log()

mpdata |> print(width=180)

p1 <- mpdata |> 
    mp_plot_rarecurve(
        .rare = RareAbundanceRarecurve,
        .alpha = Observe,
    )

# The Rarefaction of samples 
p1

# dev.off()


## calculate alpha index and visualization --------


mpdata %<>% 
    mp_cal_alpha(.abundance=RareAbundance) # Î±å¤šæ§˜æ€§æŒ‡æ•°ã®ç®—å‡º
mpdata

f1 <- mpdata  |>  
    mp_plot_alpha(
        .group=dps, 
        .alpha=c(Observe, Chao1, ACE, Shannon, Simpson, Pielou)
    ) +
    scale_fill_manual(values=c("#00A087FF", "#3C5488FF","#cab2d6"), guide="none") +
    scale_color_manual(values=c("#00A087FF", "#3C5488FF","#cab2d6"), guide="none")

f2 <- mpdata |>
    mp_plot_alpha(
        .alpha=c(Observe, Chao1, ACE, Shannon, Simpson, Pielou)
    )

# The alpha diversity comparison
f1 / f2



# é–¢æ•°ã®å†…éƒ¨è¡¨ç¤º
# MicrobiotaProcess::mp_plot_alpha
# showMethods(mp_plot_alpha)
# getMethod("mp_plot_alpha", signature = "MPSE")
# ggsignif::geom_signif()ã§ã€æ¤œå®šæ‰‹æ³•ã‚’å®šã‚ã¦ã„ã‚‹


```

## The visualization of taxonomy abundance

```{r}
#| label: The visualization of taxonomy abundance
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center


mpdata %<>%
    mp_cal_abundance( # for each samples
        .abundance = RareAbundance
    ) |>
    mp_cal_abundance( # for each groups 
        .abundance=RareAbundance,
        .group=dps
    )
mpdata

# ãƒ©ãƒ³ãƒ€ãƒ ã‚µãƒ³ãƒ—ãƒªãƒ³ã‚°ã•ã‚ŒãŸRareAbundanceã‚’ç”¨ã„ã¦ã„ã‚‹ã“ã¨ã«æ³¨æ„ï¼
p1 <- mpdata |>
    mp_plot_abundance(
        .abundance=RareAbundance,
        .group=dps, 
        taxa.class = Phylum, 
        topn = 20,
        relative = TRUE
    )
# visualize the abundance (rarefied) of top 20 phyla for each sample
p2 <- mpdata |>
    mp_plot_abundance(
        .abundance=RareAbundance,
        .group=dps,
        taxa.class = Phylum,
        topn = 20,
        relative = FALSE
    )

# The relative abundunce and abundunce of phyla of all samples
p1
p2

# The relative abundance and abundance of phyla of all samples --------------------


h1 <- mpdata |>
    mp_plot_abundance(
        .abundance = RareAbundance,
        .group = dps,
        taxa.class = Phylum,
        relative = TRUE,
        topn = 20,
        geom = 'heatmap',
        features.dist = 'euclidean',
        features.hclust = 'average',
        sample.dist = 'bray',
        sample.hclust = 'average'
    ) 
h2 <- mpdata |>
    mp_plot_abundance(
        .abundance = RareAbundance,
        .group = dps,
        taxa.class = Phylum,
        relative = FALSE,
        topn = 20,
        geom = 'heatmap',
        features.dist = 'euclidean',
        features.hclust = 'average',
        sample.dist = 'bray',
        sample.hclust = 'average'
    )
# the character (scale or theme) of figure can be adjusted by set_scale_theme
# refer to the mp_plot_dist

# The relative abundunce and abundunce of phyla of all samples
aplot::plot_list(gglist=list(h1, h2), tag_levels="A")



# visualize the relative abundance of top 20 phyla for each .group (time) --------------------

p3 <- mpdata |>
    mp_plot_abundance(
        .abundance=RareAbundance, 
        .group=dps,
        taxa.class = Phylum,
        topn = 20,
        plot.group = TRUE
    )

# visualize the abundance of top 20 phyla for each .group (time)
p4 <- mpdata |>
    mp_plot_abundance(
        .abundance=RareAbundance,
        .group= dps,
        taxa.class = Phylum,
        topn = 20,
        relative = FALSE,
        plot.group = TRUE
    )

p3
p4


```

## Beta diversity analysis

```{r}
#| label: Beta diversity analysis
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center


# standardization
# mp_decostand wraps the decostand of vegan, which provides 
# many standardization methods for community ecology.
# default is hellinger, then the abundance processed will
# be stored to the assays slot. 
mpdata %<>% 
    mp_decostand(.abundance=Abundance)
mpdata

# calculate the distance between the samples.
# the distance will be generated a nested tibble and added to the
# colData slot.
mpdata %<>% mp_cal_dist(.abundance=hellinger, distmethod="bray")
mpdata

detach("package:phyloseq", unload=TRUE)


p5 <- mpdata |> mp_plot_dist(.distmethod = bray)
p5



# when .group is provided, the dot heatmap plot with group information will be return.
p6 <- mpdata |> MicrobiotaProcess::mp_plot_dist(.distmethod = bray, .group = dps)
# The scale or theme of dot heatmap plot can be adjusted using set_scale_theme function.
p6 |> set_scale_theme(
    x = scale_fill_manual(
        values=c("orange","deepskyblue","#1b9e77"), 
        guide = guide_legend(
            keywidth = 1, 
            keyheight = 0.5, 
            title.theme = element_text(size=8),
            label.theme = element_text(size=6))), 
    aes_var = dps) |> # specific the name of variable 
    set_scale_theme(
        x = scale_color_gradient(
            guide = guide_legend(keywidth = 0.5, keyheight = 0.5)),
        aes_var = bray) |> 
    set_scale_theme(
        x = scale_size_continuous(
            range = c(0.1, 3),
            guide = guide_legend(keywidth = 0.5, keyheight = 0.5)),
        aes_var = bray)


# when .group is provided and group.test is TRUE, the comparison of different groups will be returned
p7 <- mpdata |> mp_plot_dist(.distmethod = bray, .group = dps, group.test=TRUE, textsize=2)
p7


```

## The PCoA analysis

```{r}
#| label: The PCoA analysis 
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center


library(ggplot2)

mpdata %<>% 
    mp_cal_pcoa(.abundance = hellinger, distmethod="bray")

# The dimensions of ordination analysis will be added the colData slot (default).
mpdata

# We also can perform adonis or anosim to check whether it is significant to the dissimilarities of groups.
mpdata %<>%
    mp_adonis(.abundance = hellinger, .formula = ~dps, distmethod = "bray", permutations = 9999, action = "add")
mpdata %>% mp_extract_internal_attr(name = adonis)


p8 <- mpdata %>%
    mp_plot_ord(
        .ord = pcoa, 
        .group = dps, 
        .color = dps, 
        .size = 1.8,
        .alpha = 1,
        ellipse = TRUE,
        show.legend = FALSE # don't display the legend of stat_ellipse
    ) +
    scale_fill_manual(values=c("#00A087FF", "#3C5488FF", "orange")) +
    scale_color_manual(values=c("#00A087FF", "#3C5488FF", "orange")) 

# The size of point also can be mapped to other variables such as Observe, or Shannon 
# Then the alpha diversity and beta diversity will be displayed simultaneously.
p9 <- mpdata %>% 
    mp_plot_ord(
        .ord = pcoa, 
        .group = dps, 
        .color = dps, 
        .size = Shannon,
        ellipse = TRUE,
        show.legend = FALSE # don't display the legend of stat_ellipse 
    ) +
    scale_fill_manual(
        values = c("#00A087FF", "#3C5488FF","orange"), 
        guide = guide_legend(keywidth=0.6, keyheight=0.6, label.theme=element_text(size=6.5))
    ) +
    scale_color_manual(
        values=c("#00A087FF", "#3C5488FF", "orange"),
        guide = guide_legend(keywidth=0.6, keyheight=0.6, label.theme=element_text(size=6.5))
    ) +
    scale_size_continuous(
        range=c(0.5, 3),
        guide = guide_legend(keywidth=0.6, keyheight=0.6, label.theme=element_text(size=6.5))
    )


p8 + p9

```

## Hierarchical cluster analysis

```{r}
#| label: Hierarchical cluster analysis
#| eval: false
#| Include: false


mpdata %<>%
    mp_cal_clust(
        .abundance = hellinger, 
        distmethod = "bray",
        hclustmethod = "average", # (UPGAE)
        action = "add" # action is used to control which result will be returned
    )

mpdata



# if action = 'add', the result of hierarchical cluster will be added to the MPSE object
# mp_extract_internal_attr can extract it. It is a treedata object, so it can be visualized
# by ggtree.
sample.clust <- mpdata |>
    mp_extract_internal_attr(name='SampleClust')
sample.clust


library(ggtree)
library(ggtreeExtra)
library(ggplot2)


p10 <- ggtree(sample.clust) + 
    geom_tippoint(aes(color=dps)) +
    geom_tiplab(as_ylab = TRUE) +
    ggplot2::scale_x_continuous(expand=c(0, 0.01))


p10


phyla.tb <- mpdata |> 
    mp_extract_abundance(taxa.class = Phylum, topn = 30)

# The abundance of each samples is nested, it can be flatted using the unnest of tidyr.
phyla.tb %<>% tidyr::unnest(cols=RareAbundanceBySample) |> dplyr::rename(Phyla="label")
phyla.tb


p11 <- p10 + 
    geom_fruit(
        data = phyla.tb,
        geom = geom_col,
        mapping = aes(x = RelRareAbundanceBySample, 
                      y = Sample.Name,
                      fill = label
        ),
        orientation = "y",
        #offset = 0.4,
        pwidth = 3, 
        axis.params = list(axis = "x", 
                           title = "The relative abundance of phyla (%)",
                           title.size = 4,
                           text.size = 2, 
                           vjust = 1),
        grid.params = list()
    )


p11


```

## Biomarker discovery

### Transform Abundance value:

+------------------+------------------------------------------------------------------------------+-----------------------------------+
| å¼•æ•°             | æ„å‘³                                                                         | ä¾‹                                |
+==================+==============================================================================+===================================+
| `.data`          | `MPSE`, `tbl_mpse`, `grouped_df_mpse`                                        | `mpdata`                          |
+------------------+------------------------------------------------------------------------------+-----------------------------------+
| `.abundance`     | å­˜åœ¨é‡(colnames)\                                                            | `Abundance`                       |
+------------------+------------------------------------------------------------------------------+-----------------------------------+
| `min.abun`       | å„ã‚µãƒ³ãƒ—ãƒ«ã«ãŠã‘ã‚‹æœ€å°å­˜åœ¨é‡                                                 | `1` â†’ å„ã‚µãƒ³ãƒ—ãƒ«ã§1ä»¥ä¸Š           |
+------------------+------------------------------------------------------------------------------+-----------------------------------+
| `min.prop`       | TaxaãŒå­˜åœ¨ã™ã¹ã æœ€å°ã‚µãƒ³ãƒ—ãƒ«å‰²åˆï¼ˆ0â€“1ï¼‰ã¾ãŸã¯æ•°å€¤ï¼ˆæ•´æ•°ï¼‰â†’ Precalenceã®ã“ã¨ | `0.1` â†’ å…¨ä½“ã®10%ã®ã‚µãƒ³ãƒ—ãƒ«ã«å­˜åœ¨ |
+------------------+------------------------------------------------------------------------------+-----------------------------------+
| `include.lowest` | `TRUE` ã®ã¨ã `>= min.abun`ã€`FALSE` ãªã‚‰ `> min.abun`ï¼ˆãƒ‡ãƒ•ã‚©ãƒ«ãƒˆï¼‰         | `TRUE` ãªã‚‰é–¾å€¤å«ã‚€               |
+------------------+------------------------------------------------------------------------------+-----------------------------------+

```{r}
#| label: Transform Abundance value
#| eval: false
#| Include: false

load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/SaveObjects/mpdata.RData")

mpdata %<>%
    mp_cal_abundance(.abundance = Abundance, action = "add",
                     relative = TRUE, force = TRUE)
mpdata %<>%
    mp_filter_taxa(.abundance = RelAbundanceBySample,
                   min.prop = 0.12,
                   min.abun = 0.1 # ç›¸å¯¾å­˜åœ¨é‡ 0.1% > Filtering  
    )

### Differentially abundant taxa --------------

mpdata %<>%
    mp_diff_analysis(
       .abundance = Abundance, # å­˜åœ¨é‡ã‚’æŒ‡å®š (ç›¸å¯¾å­˜åœ¨é‡ or ãƒ©ãƒ³ãƒ€ãƒ ã‚µãƒ³ãƒ—ãƒªãƒ³ã‚°å­˜åœ¨é‡ or çµ¶å¯¾å­˜åœ¨é‡ or æ­£è¦åŒ–å­˜åœ¨é‡)
       .group = dps, # æ¯”è¼ƒå¯¾è±¡
       first.test.alpha = 0.3, # æœ‰æ„æ€§ã®ä¸€æ¬¡åˆ¤å®šã«ç”¨ã„ã‚‹på€¤ã®ã—ãã„å€¤
       second.test.alpha = 0.05, action = "add", relative = TRUE # å¤šé‡æ¤œå®šè£œæ­£å¾Œã®ã—ãã„å€¤
    )

```

`mp_diff_analysis()`ï¼šæ¯”è¼ƒã—ãŸã„å‡¦ç†åŒºã«ã¯ã€æœ€ä½ã§ã‚‚ 5 Sampleã¯å¿…è¦

# DESeq2

[Analyzing RNA-seq data with DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)\
[RNA-seq workflow: gene-level exploratory analysis and differential expression](https://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html)\
[DESeq2 with phyloseq](https://joey711.github.io/phyloseq-extensions/DESeq2.html)

## Preparing quantification input to DESeq2

The values in the matrix should be counts or estimated counts of sequencing reads/fragments\
This is important for `DESeq2`â€™s statistical model to hold, as only counts allow assessing the measurement precision correctly\
It is important to never provide **counts that were pre-normalized for sequencing depth/library size, as the statistical model is most powerful when applied to un-normalized counts**, and is designed to account for library size differences internally\
The object class used by the DESeq2 package to store the read counts and the intermediate estimated quantities during statistical analysis is the **DESeqDataSet**, which will usually be represented in the code here as an object `dds`\
[We will now show 4 ways of constructing a **DESeqDataSet**, depending on what pipeline was used upstream of DESeq2 to generated counts or estimated counts:]{.mark-blue}

-   From transcript abundance files and tximport
-   **From a count matrix**
-   From htseq-count files
-   From a SummarizedExperiment object

Alternatively, the function `DESeqDataSetFromMatrix` can be used if you already have a matrix of read counts prepared from another source

## PhyseqObjectsDESeq2

```{r}
#| label: PhyseqObjectsDESeq2
#| eval: false
#| Include: false

rm(list = ls())
library("phyloseq")
library("DESeq2")

load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/SaveObjects/PhyloseqData.RData")

PhyseqData <- transform_sample_counts(PhyseqData, function(x) 100*(x / sum(x)))
# PhyseqData <- filter_taxa(PhyseqData, function(x) mean(x) > 0.1, TRUE)

# write.csv(t(PhyseqData@otu_table@.Data),
#            file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/PhyseqData@otu_table@.Data.csv")


dps_dds = phyloseq_to_deseq2(PhyseqData, ~ dps)
dps_dds = DESeq(dps_dds, test="Wald", fitType="parametric")
res = results(dps_dds, cooksCutoff = FALSE)
alpha = 0.01
# write.csv(res, file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/res.csv")
sigtab = res[which(res$padj < alpha), ] 
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(PhyseqData)[rownames(sigtab), ], "matrix"))
head(sigtab)



```

### Taxa Level in Family Level:

[**æ³¨æ„**]{.mark-red}

tax_glom(taxrank = "Family") â†’ ASVsã®åˆ†é¡ç¾¤ã‚’çµ±åˆã—ã¦ã„ã‚‹ãŸã‚ã€ç§‘(Family)ä»¥ä¸‹ã®åˆ†é¡ç¾¤ã§ã‚ã‚‹å±(Genus)ãŒç•°ãªã‚‹ASVsã‚‚ã€æ•°å­—ãŒå°ã•ã„ASVsã«æ ¼ç´ã•ã‚Œã‚‹

â†’ ASV 1 ã®FamilyãŒé¡•è‘—ã«å­˜åœ¨é‡ãŒå¤‰å‹•ã—ãŸã¨ã—ã¦ã‚‚ã€ãã®ä¸­ã®GenusãŒåŒæ§˜ã®å‚¾å‘ã‚’ç¤ºã—ã¦ã„ã‚‹ã¨ã¯é™ã‚‰ãªã„!!!!

```{r}
#| label: Taxa Level in Family Level
#| eval: true
#| Include: false
#| message: false

### Taxa Level in Family Level --------------------------------
rm(list = ls())
load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/SaveObjects/PhyloseqData.RData")
library("phyloseq")
library("DESeq2")


#### Filtering ---------------------
PhyseqData_Family <- PhyseqData  |> 
    tax_glom(taxrank = "Family") |>                  # agglomerate at phylum level
    filter_taxa(function(x) mean(x) > 100, TRUE) |>  # Readæ•°ãŒç´„100000ã§ã‚ã‚Šã€0.1%ã®Readsæ•° 
    phyloseq::subset_taxa(Kingdom == "Bacteria")               # Archaeaã‚’é™¤å»


# PhyseqData_Family.csv <- read.csv("~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/Filtering_ASVs_Table/PhyseqData_Family.csv",
#             header = TRUE, sep = ",", fileEncoding = "UTF-8")

# write.csv(cbind(t(PhyseqData_Family@otu_table@.Data), PhyseqData_Family@tax_table@.Data),
#           file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/Filtering_ASVs_Table/PhyseqData_Family.csv")




#### DESeq2 conversion & results table ----------------

dps_dds = phyloseq_to_deseq2(PhyseqData_Family, ~ dps)       # dpsåˆ—(æ•£å¸ƒå¾Œæ—¥æ•°)ã§æ¯”è¼ƒ
dps_dds = DESeq(dps_dds, test="Wald", fitType="parametric")



res = results(dps_dds, cooksCutoff = FALSE)
# write.csv(res, file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/èŠè‰/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/Filtering_ASVs_Table/res.csv")
 


alpha = 0.01
sigtab = res[which(res$padj < alpha), ] 
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(PhyseqData)[rownames(sigtab), ], "matrix"))
head(sigtab)
dim(sigtab)

res

#### Results_ggplot ----------------------------
library("ggplot2")
theme_set(theme_bw())


# Phylum order
x = tapply(sigtab$log2FoldChange, sigtab$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtab$Phylum = factor(as.character(sigtab$Phylum), levels=names(x))
# Genus order
x = tapply(sigtab$log2FoldChange, sigtab$Genus, function(x) max(x))
x = sort(x, TRUE)
sigtab$Genus = factor(as.character(sigtab$Genus), levels=names(x))

# ggplot(sigtab, aes(x=Family, y=log2FoldChange, color=Phylum)) + geom_point(size=abs(sigtab$log2FoldChange)) + 
#   theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))



sigtab$Sign <- ifelse(sigtab$log2FoldChange < 0, "Negative", "Positive")
sigtab$ASV <- rownames(sigtab)
custom_colors_21 <- c(
    "#E41A1C", "#377EB8", "#4DAF4A", "#984EA3", "#FF7F00","#FFFF33", "#A65628", "#F781BF", "#999999", "#66C2A5",
    "#FC8D62", "#8DA0CB", "#E78AC3", "#A6D854", "#FFD92F","#E5C494", "#B3B3B3", "#1B9E77", "#D95F02", "#7570B3","#E7298A"
)


ggplot(sigtab, aes(x = Family, y = abs(sigtab$log2FoldChange), color = Phylum, shape = Sign)) +
    geom_point(size = 5) +
    geom_text(aes(label = ASV), vjust = -1, size = 3) + 
    scale_color_manual(values = custom_colors_21) +
    theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.5)) +
    scale_shape_manual(values = c("Positive" = 16, "Negative" = 17)) +  # ä¸¸ã¨ä¸‰è§’ãªã©
    ylab("log2FoldChange") +
    xlab("Family") +
    theme(legend.position = "right")



```

### Taxa Level in Order Level:

```{r}
#| label: Taxa Level in Order Level
#| eval: false
#| Include: false



```

### Taxa Level in Class Level:

```{r}
#| label: Taxa Level in Class Level
#| eval: false
#| Include: false



```

### Taxa Level in Phylum Level:

```{r}
#| label: Taxa Level in Phylum Level
#| eval: false
#| Include: false



```

# SessionInfo

```{r}
#| label: SessionInfo
#| eval: true
#| Include: false


sessionInfo()


```
