---
title: "Phyloseq_RawData_Bacteria_PreProcess"
---

::: callout-note
## Reference & Memo

\[R Package\] [Scales](https://scales.r-lib.org/)

[È£üÂìÅÂæÆÁîüÁâ©Â≠¶ÔºàÊ§úÊüª„Å®Âà∂Âæ°ÊñπÊ≥ïÔºâÔΩúÂü∫Á§é„Å®ÊúÄÊñ∞ÊÉÖÂ†±„ÇíËß£Ë™¨ÔΩúÊú®Êùë„ÄÄÂá°](https://foodmicrob.com/16s-amplicon-sequencing-food-microbiology/)

[Phyloseq HP](https://joey711.github.io/phyloseq/)

<https://ushio-ecology-blog.blogspot.com/2021/04/20210428blogger0015.html>

[Bioconductor::MicrobiotaProcess](https://www.bioconductor.org/packages/release/bioc/html/MicrobiotaProcess.html)

[Introduction to MicrobiotaProcess](https://bioconductor.org/packages/devel/bioc/vignettes/MicrobiotaProcess/inst/doc//MicrobiotaProcess.html#biomarker-discovery)

[Github_MicrobiotaProcess](https://github.com/YuLab-SMU/MicrobiotaProcess)

[ggtreeExtra HP](https://bioconductor.org/packages/release/bioc/vignettes/ggtreeExtra/inst/doc/ggtreeExtra.html)

[Metagenomic biomarker discovery and explanation](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2011-12-6-r60)

[SILVA138.2 SSUrRNA database](https://www.arb-silva.de/)
:::

# NGS Analysis Flow

::: callout-note
## [Filtering]{.mark-cyan}

‰ª•‰∏ã„ÄÅË´ñÊñá„Åß‰Ωø„Çè„Çå„Å¶„ÅÑ„ÇãÊâãÊ≥ï

„ÉªÁõ∏ÂØæÂ≠òÂú®Èáè 1% \>\
„ÉªÂπ≥ÂùáÁõ∏ÂØæÂ≠òÂú®Èáè 10\^-5 \>\
„ÉªPrevalence„Åß„ÄÅsample„ÅÆÂÜÖ„ÄÅ?„Åß„Åó„ÅãÁ¢∫Ë™ç„Åï„Çå„Å¶„ÅÑ„Å™„ÅÑASVs(OTUs)„ÇíÊéíÈô§\
„ÉªÂá¶ÁêÜÂå∫(Ëæ≤Ëñ¨Êï£Â∏É or ÁóÖÂÆ≥Áô∫Áîü„Çµ„É≥„Éó„É´)„ÅßÂ≠òÂú®„ÅåÁ¢∫Ë™ç„Åï„Çå„Å™„ÅÑASVs„ÅØÊéíÈô§„Å™„Å©

‚Üí Filtering„ÅÆÂü∫Ê∫ñ„ÅØ„ÄÅË´ñÊñá„Å´„Çà„Å£„Å¶Êßò„ÄÖ„Åß„ÄÅ„Åä„Åù„Çâ„Åè„Åù„ÅÆÊôÇ„ÄÖ„ÅßÈÉΩÂêà„Åå„ÅÑ„ÅÑ„Çà„ÅÜ„Å´Ëß£Èáà??

Q: What is the difference between averaging, homogenization and standardization?

Averaging and homogenization both refer to the random sampling of sequences from different samples according to a certain rule (e.g., we go by the minimum number of sequences) so that the number of sequences in different samples is the same.

Standardization is generally a Z-value calculation for data that fluctuates over a large range, scaling the data so that it falls into a small specific interval that is easy to display in a graph.
:::

::: callout-note
## Â§öÊßòÊÄßË©ï‰æ°

‚Üí Êï∞ÂçÉ„ÅÆASVs(OTU)„ÅåÊ§úÂá∫„Åï„Çå„Çã„Åü„ÇÅ„ÄÅÂêÑASVs„Å´ÁùÄÁõÆ„Åô„Çã„Å†„Åë„Åß„Å™„Åè„ÄÅ„Çµ„É≥„Éó„É´ÂÜÖ„ÅÆÂæÆÁîüÁâ©Âè¢(ASVsÂÖ®‰Ωì)„ÇíÊßãÈÄ†„ÇÑÁµÑÊàê„Å®„ÅÑ„Å£„ÅüË¶≥ÁÇπ„Åã„Çâ„ÄÅË©ï‰æ°„Åô„Çã„Åü„ÇÅ\
„Åì„Çå„Åæ„ÅßÂÖàË°åÁ†îÁ©∂„Åß„ÅØ„ÄÅÊúâÊ©üËÇ•Êñô„ÅåÂæÆÁîüÁâ©Êï∞„ÇÑÂè¢„ÇíÊ¥ªÊÄßÂåñ„Åï„Åõ„Å¶„ÄÅÁóÖÂÆ≥Áô∫Áîü„ÇíÊäëÂà∂„Åô„Çã„Å™„Å©Â†±Âëä„Åï„Çå„Å¶„ÅÑ„Çã„Åå„ÄÅ„ÅÇ„ÇãÁâπÂÆö„ÅÆÂæÆÁîüÁâ©(Á¥∞Ëèå„ÉªÁ≥∏Áä∂Ëèå)„ÅåÂØÜÊé•„Å´ÁóÖÂÆ≥Áô∫Áîü„Å´Èñ¢„Çè„Å£„Å¶„ÅÑ„ÇãÂèØËÉΩÊÄß„ÇÇ„ÅÇ„Çã„Åå„ÄÅ„Åù„Çå„Çà„Çä„ÇÇË§áÊï∞„ÅÆÂæÆÁîüÁâ©ÂêåÂ£´„ÅåË§áÈõë„Å´Èñ¢„Çè„ÇäÂêà„Å£„Å¶„ÄÅÁóÖÂéüËèå„ÅÆÊÑüÊüì„Åå„ÄÅÁ´∂ÂêàÁöÑ„Å´Âä£„Çä„ÄÅÁóÖÂÆ≥Áô∫Áîü„Å´Ëá≥„Å£„Å¶„ÅÑ„Å™„ÅÑÂèØËÉΩÊÄß„ÇÇ„ÅÇ„Çã„Åü„ÇÅ\
„Åì„ÅÆË§áÈõëÊÄß„ÇíÁ∞°Âçò„Å´Áµ±Ë®àÁöÑ„Å´„ÇÇÁßëÂ≠¶ÁöÑ„Å´„ÇÇË©ï‰æ°„Åô„Çã„ÅÆ„ÅØ„ÄÅ‰∏çÂèØËÉΩ„Å´Ëøë„ÅÑ„Åì„Å®„ÇíÂâçÊèê„Å´ÁΩÆ„ÅÑ„Åü‰∏ä„Åß„ÄÅÂæÆÁîüÁâ©Âè¢Ëß£Êûê„ÅÆË´ñÊñá„Å™„Å©„ÇíË™≠„ÇÄÊñπ„ÅåËâØ„ÅÑ„ÄÇËß£Êûê„ÅØ„ÄÅËëóËÄÖ„ÅÆ‰∏ªË¶≥„ÇÇÂÖ•„Çã„Åì„Å®„ÇÇ„ÅÇ„Çä„ÄÅÊâπÂà§ÁöÑ„Å™Ë¶ñÁÇπ„ÅßË™≠„ÇÄ„Åì„Å®

„ÉªŒ±Â§öÊßòÊÄß(Sobs, Chao1, Shannon, and Faith‚Äôs PD)\
‚Üí ANOVA & Tukey‚Äôs HSD post hoc analyses

„ÉªŒ≤Â§öÊßòÊÄß\
The Shannon index curves & Rarefaction curves\
beta diversity as represented by Hellinger transformed OTU relative abundances and weighted UniFrac distances\
‚Üí Tukey‚Äôs HSD post hoc analyses\

Differences in fungal and prokaryotic community composition were tested through permutational multivariate analysis of variance (PERMANOVA) with the ‚Äòadonis2‚Äô function on Bray-Curtis distances in the R package vegan \[45\]

„ÉªASVsÊï∞„Å´Âü∫„Å•„Åè„ÄÅÁæ§ÈõÜÁµÑÊàê„ÅÆË©ï‰æ°

„Éª
:::

::: callout-note
## Áæ§ÈõÜÊßãÈÄ†„ÅÆË©ï‰æ° (community composition)

„ÉªPCoA ‚Üí Ë™øÊüª„Åó„ÅüSampleÊï∞(ÂúüÂ£å„Çµ„É≥„Éó„É´(eDNA))„ÅåÂ§ö„ÅÑÊôÇ„Å´„ÅØ„ÄÅÈùûÂ∏∏„Å´ÊúâÂäπ„Å†„Åå„ÄÅ„Çµ„É≥„Éó„É´Êï∞„ÅåÂ∞ë„Å™„ÅÑÂ†¥Âêà„ÅØ„ÄÅËß£Êûê„ÅÆÂâçÊèê„ÅåÊàêÁ´ã„Åó„Å™„ÅÑ„Åì„Å®„ÇÇ„ÅÇ„Çã„ÅÆ„ÅßÊ≥®ÊÑè

„ÉªNMDS(ANOSIM\[vegan\]„ÅßÊúâÊÑèÂ∑ÆÊ§úÂÆö
:::

::: callout-note
## Differentially abundant taxa

[**ÊúâÊÑè„Å´Â≠òÂú®Èáè„ÅåÁï∞„Å™„ÇãÂàÜÈ°ûÁæ§(ASVs)„ÅÆÊé¢Á¥¢**]{.mark-cyan}

„ÉªÁõ∏ÂØæÂ≠òÂú®Èáè„ÅÆÊØîËºÉ(one-way ANOVA, LSD)\
‚Üí ‰∏ªË¶Å„Å™ÂàÜÈ°ûÁæ§„ÅÆÁâπÂæ¥„ÇíÊé¢Á¥¢

„Éª„É©„É≥„ÉÄ„É†„Éï„Ç©„É¨„Çπ„Éà„É¢„Éá„É´„Åß„ÄÅËæ≤Ëñ¨Êï£Â∏É„Å´„Çà„Å£„Å¶Â§âÂãï„Åó„ÅüÂæÆÁîüÁâ©Âè¢„Å´ÂèçÂøú„Åó„ÅüÂàÜÈ°ûÁæ§(ASVs)„Çí‰∫àÊ∏¨\
We built random forest regression models to test the effect of altered prokaryote abundance through fungal diversity by using prokaryote abundances to predict fungal diversity.

Random forest models were generated with the‚ÄòrandomForest‚Äô function in the randomForest R package \[49\]. To remove redundant features and avoid over tting models, we removed redundant OTUs with the‚ÄòBoruta‚Äô function in the package Boruta \[50\].

The method performs a top-down search for relevant OTUs by comparing the importance of the original OTUs from those selected at random. Models were tuned to achieve the lowest stable out-of-bag (OOB) error estimate possible, and the best mtry value (number of OTUs sampled at random in the entire pool for each tree at each split) was selected using the‚ÄòtuneRF‚Äô function in randomForest R package.

„ÉªFungal OTUs were determined to be signicant if the W value was greater than 70% of the taxa tested based on Wilcoxon ranked sum test between additive log-ratio transformed data and a Benjamini- Hochbergj adjusted P value (Œ± = 0.05) \[47\].

„ÉªANCOMÔºàAnalysis of Composition of MicrobiomesÔºâ

‚úÖ ÂÆöÁæ© ANCOM„Å®„ÅØ„ÄÅÂæÆÁîüÁâ©Áæ§ÈõÜ„Å´„Åä„Åë„ÇãÁæ§Èñì„ÅÆÂ∑ÆÁï∞Ôºà‰æã„Åà„Å∞Âá¶ÁêÜÂå∫„Å®ÂØæÁÖßÂå∫„Å™„Å©Ôºâ„Çí„ÄÅÁõ∏ÂØæÁöÑ„Å™„Éá„Éº„ÇøÔºàcompositional dataÔºâ„Å´Âü∫„Å•„ÅÑ„Å¶Ëß£Êûê„Åô„ÇãÊñπÊ≥ï„Åß„Åô„ÄÇ Áâπ„Å´„ÄÅ16S rRNA„Ç∑„Éº„Ç±„É≥„Çπ„Å™„Å©„ÅÆ„Éá„Éº„Çø„Åß„ÅØ„ÄÅÁ∑èÈáèÔºàlibrary sizeÔºâ„Å´Âà∂Èôê„Åå„ÅÇ„Çã„Åü„ÇÅ„ÄÅÂæó„Çâ„Çå„Çã„Éá„Éº„Çø„ÅØ„ÄåÁõ∏ÂØæÁöÑ„Äç„Å™„ÇÇ„ÅÆ„Åß„ÅÇ„Çã„Åì„Å®„Åã„Çâ„ÄÅÈÄöÂ∏∏„ÅÆÁµ±Ë®àËß£Êûê„Åß„ÅØË™§Â∑Æ„ÇÑ„Éê„Ç§„Ç¢„Çπ„ÅåÁîü„Åò„ÇÑ„Åô„Åè„Å™„Çä„Åæ„Åô„ÄÇ\
ANCOM„ÅØ„Åù„ÅÆÁÇπ„ÇíËÄÉÊÖÆ„Åó„ÄÅÁõ∏ÂØæ„Éá„Éº„Çø„Å´ÈÅ©„Åó„ÅüÊñπÊ≥ï„Åß„Äå„Å©„ÅÆÂàÜÈ°ûÁæ§ÔºàËèåÁ®Æ„ÇÑÂ±û„Å™„Å©Ôºâ„ÅåÊúâÊÑè„Å´Áï∞„Å™„Çã„Åã„Äç„ÇíÊ§úÂá∫„Åô„ÇãÊâãÊ≥ï„Å®„Åó„Å¶ÈñãÁô∫„Åï„Çå„Åæ„Åó„Åü„ÄÇ

BioManeger::ANCOMBS package

„ÉªLEfSeËß£Êûê ‚Üí ÊØîËºÉ„Åó„Åü„ÅÑÂá¶ÁêÜÂå∫„Åß„ÄÅÊúÄ‰Ωé„Åß„ÇÇ5„Å§‰ª•‰∏ä„ÅØ„Çµ„É≥„Éó„É´„ÅåÂøÖË¶Å„Å™„ÅÆ„Åß„ÄÅÊ≥®ÊÑè!!!
:::

::: callout-note
## ASVs„ÅÆÁâπÂæ¥„ÇÑÊÄßË≥™Ë™øÊüª

LPSNÔºö[List of Prokaryotic names with Standing in Nomenclature](https://lpsn.dsmz.de/text/introduction)

‚Üí ÊúâÊÑè„Å´Â§âÂãï„Åó„ÅüASVs„ÅÆÁßë„ÉªÂ±û„Å™„Å©„ÇíÊ§úÁ¥¢„Åó„ÄÅ„Åù„ÅÆASVs„Å´Èñ¢„Åô„ÇãË´ñÊñá„ÇíËæø„Çã

GTDBÔºö[Genome Taxonomy Database](https://gtdb.ecogenomic.org/)

‚Üí ???

IMGÔºö[Integrated Microbial Genomes](https://img.jgi.doe.gov/)

‚Üí ???

PubMedÔºö???\[\] AND ??„ÅßÊ§úÁ¥¢„Åô„Çã

Bryobacter\[Title/Abstract\] AND 16S
:::

# Paper Information

::: callout-note
## [SILVA Release 138.2](https://www.arb-silva.de/documentation/release-1382/)

The SILVA taxonomy release 138.2 is a maintenance release, with no additional sequences added to the ARB databases since the previous release 138.1. In this update we have focused on Prokaryotes.

Key changes:

-   bug fixes

-   extensive adoption of the LPSN taxonomy at the genus and phylum level, synchronised by April 2023

-   ongoing adoption of the LPSN taxonomy for families, orders and classes

-   a standard six-rank taxonomy has been established for all prokaryotic sequences

-   **the term "uncultured" to refer to clades with unassigned taxonomy has been discontinued**

-   **widespread adoption of the term "Incertae Sedis" to refer to unassigned taxa**

For each SILVA release, the taxonomy has to be frozen at a certain point in time.

The reason for freezing the taxonomy for this release so early is that the migration of SILVA from the Max Planck Institute for Marine Microbiology to the DSMZ caused more problems than originally expected.

The release was therefore delayed considerably. However, these problems also had a positive outcome: by adding the de.NBI cloud as a computing instance, the redundancy for the SILVA website jobs (alignment, TestProbe, TestPrime, export) could be increased.

The release required not only changes to the SILVA pipeline software, but also adjustments to the DSMZ infrastructure.

Normally, our long-time taxonomic curator would have used the delay for further updates.

Unfortunately, he has been unavailable since last summer due to a career change. After two years of migration and major changes in the team, we are now on the verge of filling all positions and returning to normal operations.

Although the situation has not been easy for us or our users, we are very optimistic about the future now that SILVA is permanently funded.

We plan to release one full release and one taxonomic update per year, starting next year at the latest. We would like to thank you for your loyalty over the years and take this as an incentive to continue to provide the high quality you have come to expect.
:::

::: callout-note
## Waste not, want not: revisiting the analysis that called into question the practice of rarefaction

they removed rare and low-prevalence OTUs in two steps.

First, they removed any OTUs whose total abundance was less than three across all 80 samples and that did not appear in at least three samples.

Second, they removed any OTUs that did not have more than one sequence in more than 5% of the 80 samples (i.e., four samples) and that did not have a total abundance across the 80 samples greater than one half of the number of samples in each community type (i.e., 20)

Relative abundance data were used to calculate Bray-Curtis, Unweighted UniFrac, and Weighted UniFrac distances

„ÄêDiscussion„Äë

It is worth commenting on WNWN‚Äôs advice to use DESeq‚Äôs Variance Stabilization or edgeR‚Äôs Upper-Quartile Log-fold Change normalization strategies. These methods have been adopted from gene expression analysis to microbiome analysis.

Gene expression analysis implicitly assumes that all samples have the same genes. While this assumption might be valid when comparing host gene expression in healthy and diseased tissues from a cohort of patients, it does not generalize to those patients‚Äô microbiota.

Microbial populations are highly patchy in their distribution. Thus, a zero count for gene expression is more likely to represent a gene below the limit of detection, whereas a zero count for a microbiome analysis is more likely to represent the true absence of the OTU.

In fact, zeroes are less common in gene expression analyses. However, because microbiome studies have so many zeroes, it is necessary to add a pseudo-count to all OTUs for both the edgeR- and DESeq-based normalizations (L443 and L487, respectively).

In WNWN, a pseudo-count of 1 was used. However, this value is arbitrary and the results can vary based on the choice of pseudo-count and the patchiness of the communities being analyzed.

Since WNWN was published, compositional approaches have been proposed to account for uneven sequencing and to provide improved interpretability (14, 16‚Äì18, 44, 45). However, these methods also often require the use of pseudo-counts. Rarefaction is preferred to these alternatives.

[The choice of distance metric is a complicated question]{.mark-cyan} ‚Üí Note_KeyNotes (SCISPACE)„ÇíÂèÇÁÖß ‚Üí Rarefaction„Åå„ÄÅ„Çµ„É≥„Éó„É´„ÅßÁï∞„Å™„Çã„Ç∑„Éº„Ç±„É≥„ÇπÊ∑±Â∫¶„ÅÆÂΩ±Èüø„ÇíÂà∂Âæ°„Åô„ÇãÂîØ‰∏Ä„ÅÆÊñπÊ≥ï„Åß„ÅÇ„Çã„Åã„ÇÇ„Åó„Çå„Å™„ÅÑ...„ÄÅÁõ∏ÂØæÂ≠òÂú®Èáè„ÅÆÂà©Áî®„Å´„ÇÇÊ≥®ÊÑè„ÅåÂøÖË¶Å ‚Üí Reference„ÅÆË´ñÊñá„ÇÇË™≠„Çì„Åß„Åø„Çã
:::

::: callout-note
## [Non-target impacts of fungicide disturbance on phyllosphere yeasts in conventional and no-till management](https://www.nature.com/articles/s43705-022-00103-w#Sec15)

**\[Suppliments\]**

Import and Preprocessing in R. Data were imported into R 4.0.3 \[17\], and the R packages phyloseq 1.24.2 \[18\] and vegan 2.5.3 \[19\] were used for most analyses.

**Samples with low sequencing coverage (less than 1000 reads) were removed from the analysis.**

Contaminant OTUs (i.e., those prevalent in negative extraction controls) were removed with the R package decontam \[20\].

Before normalization, richness was assessed for Prokaryotes and Fungi in the leaves and roots of each crop using the ‚Äòestimate_richness‚Äô function in the phyloseq package.

Results of alpha diversity analyses were plotted using the ‚Äòggplot2‚Äô package \[21\].

Then, sample read counts were normalized using the cumulative sum scaling technique within the metagenomeSeq R package \[22\].

**\[Statistical analysis:\]**

For differential abundance analysis, fungal OTUs (fOTU, hereafter) with a [**mean relative abundance less than 10\^‚àí5**]{.underline} and [**fOTUs with zeros present in 95% samples**]{.underline} were discarded from the analysis to avoid detecting fOTUs as signi cantly different based on stochasticity.\
[**In addition, fOTUs that were never present in fungicide treated plots were not included.**]{.underline}\
‚Üí Prevalence + Áõ∏ÂØæÂ≠òÂú®Èáè„ÅßFiltering + Ëæ≤Ëñ¨Âá¶ÁêÜÂå∫„ÅßÂ≠òÂú®„Åó„Å¶„ÅÑ„Å™„ÅÑOTUs„ÅØÈô§Â§ñ„Åó„Å¶„ÄÅÂÆüÊñΩ
:::

::: callout-note
## MicrobiotaProcess: A comprehensive R package for deep mining microbiome
:::

::: callout-note
## Non-target impacts of fungicide disturbance on phyllosphere yeasts in conventional and no-till management

\[Statistical analysis:\]

Differences in fungal and prokaryotic community composition were tested through permutational multivariate analysis of variance (PERMANOVA) with the ‚Äòadonis2‚Äô function on Bray-Curtis distances in the R package vegan

Differentially abundant taxa resulting from fungicide application were identi ed by comparing fungicide treated plots to control plots through an analysis of the composition of microbiomes (ANCOM v 2.1)\
**‚Üí Âá¶ÁêÜÂå∫„Å´ÂøÖË¶Å„Å™„Çµ„É≥„Éó„É´Êï∞„Åå5„Å§‰ª•‰∏ä**\
„ÄêAnalysis of composition of microbiomes: a novel method for studying microbial composition„Äë
:::

::: callout-note
## The core bacterial microbiome of banana (Musa spp.)

The main and interactive effects of compartment and soil/genotype on the composition of bacterial communities (i.e. beta diversity as represented by Hellinger transformed OTU relative abundances and weighted UniFrac distances) were assessed using Permutational Multivariate Analy- sis of Variance (PERMANOVA) \[41\] as implemented in vegan \[42\].

Defining ‚Äòcandidate core‚Äô bacterial taxa For each level of a significant factor (i.e. soil or genotype), we selected can- didate-core OTUs on the basis that they were in ‚â• 50% of replicates within one or more compartments, at a mean relative abundance of ‚â• 0.5% where present.
:::

::: callout-note
## Role of the beneficial phyllosphere microbiome in the defense against red clover anthracnose caused by Colletotrichum americae-borealis

Statistical analysis

Use Shapiro-Wilk test and Anderson-Darling test to check for significant deviations from normal distribution.

The disease index data were plotted as a histogram with an overlaid normal distribution curve using ggplot2 package in R.

The Kruskal‚ÄìWallis test was used to assess significant differences in alpha diversity among groups. PERMANOVA was used to examine the significant differences in microbial communities (i. e., beta diversity) across the sample groups.

The Wilcoxon test was used to determine the statistical significance of differences in relative abundance among taxonomic groups.

Duncan‚Äôs new multiple range test was used for significance analysis of the data, and the results are expressed as the means ¬±standard deviations.
:::

# Chunk Options & SetDirectory

```{r}
#| label: ChunkOption
#| include: true
#| message: false

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome")

# https://astrobiomike.github.io/amplicon/dada2_workflow_ex

```

# MergeingMethods:

::: callout-note
MergingMethods

Merging methods include `merge_taxa` and `merge_samples`, intended for merging specific OTUs or samples, respectively.

There is also the `merge_phyloseq` function for a complete merge of two or more phyloseq-objects (or a phyloseq-object and one or more separate components). For example, the following code merges the first 5 OTUs in the Chlamydiae-only dataset.

**ÂàÜÈ°û„É¨„Éô„É´„ÅÆÁµ±‰∏Ä**

OTU(or ASVs)„ÅØÈÄöÂ∏∏„ÄÅ**„Åß„Åç„Çã„Å†„ÅëÁ¥∞„Åã„ÅÑÂàÜÈ°ûÔºàÁ®Æ„É¨„Éô„É´„Å™„Å©Ôºâ** „ÅßÂàÜ„Åë„Çâ„Çå„Çã„Åå„ÄÅÂ†¥Âêà„Å´„Çà„Å£„Å¶„ÄÅ**‰∏ä‰ΩçÂàÜÈ°ûÔºàÂ±û„É¨„Éô„É´„ÇÑÁßë„É¨„Éô„É´Ôºâ„ÅßËß£Êûê„ÇíË°å„ÅÜÊñπ„ÅåÈÅ©„Åó„Å¶„ÅÑ„Çã** „Åì„Å®„Åå„ÅÇ„Çã„ÄÇ

> Â±û„ÇÑÁßë„Åß„ÄÅRelativeAbundunce„Çíplot„Åô„ÇãÂ†¥Âêà„Å™„Å©„Å´Âà©Áî®„Åô„Çã

‚úÖ **ÁâπÂÆö„ÅÆÂàÜÈ°û„É¨„Éô„É´Ôºà‰æã: Â±û„É¨„Éô„É´Ôºâ„Åß„ÅÆËß£Êûê„ÇíË°å„ÅÜ„Åü„ÇÅ„Å´ OTU „ÇíÁµ±Âêà„Åô„Çã**

-   ‰æã: **Âêå„ÅòÂ±û (`Genus`) „Å´Â±û„Åô„Çã OTU „Çí„Åæ„Å®„ÇÅ„Çã**

-   ‰æã: **ÈñÄ (`Phylum`) „ÇÑ„ÇØ„É©„Çπ (`Class`) „Åß„Éá„Éº„Çø„ÇíÊï¥ÁêÜ„Åô„Çã**

> **üîπ ‰æã„Åà„Å∞**
>
> -   `OTU_1` ‚Üí *Bacteroides fragilis*
>
> -   `OTU_2` ‚Üí *Bacteroides vulgatus*
>
>     ‚Üí **„Åì„Çå„Çâ„Çí "Bacteroides Â±û" „Å®„Åó„Å¶Áµ±Âêà„Åô„Çã„ÄÅ„Å§„Åæ„Çä‰∏ä‰Ωç„ÅÆÂàÜÈ°û„É¨„Éô„É´„ÅßÊØîËºÉ„Åó„Åü„ÅÑÂ†¥Âêà„Å´„ÄÅ‰ΩøÁî®„Åô„Çã„ÄÅÁâπÂÆö„ÅÆASVs„ÇíÁµ±Âêà„Åô„ÇãÂ†¥Âêà„Å´‰ΩøÁî®„Åô„Çã**
>
>     ‚Üí ÂÖ®„Å¶„ÅÆÂ±û„ÇíÁµ±Âêà„Åô„Çã„Å™„Çâ„ÄÅ‰ª•‰∏ã„ÅÆ„Çà„ÅÜ„Å´ÂàÜÈ°ûÂçò‰Ωç„ÇíÊåáÂÆö„Åô„Çå„Å∞ËâØ„ÅÑ\
>     `tax_glom(physeq„Ç™„Éñ„Ç∏„Çß„ÇØ„Éà, taxrank = "Family", NArm = true/false)`
>
>     ‚Üí **Áµ±ÂêàÂæå„ÅØ„ÄÅÁµ±Âêà„Åó„ÅüÂàÜÈ°û„É¨„Éô„É´‰ª•Èôç„ÅØ„ÄÅNA„ÅßË°®Á§∫„Åï„Çå„Çã**

------------------------------------------------------------------------

data("GlobalPatterns")

GP.chl = subset_taxa(GlobalPatterns, Phylum=="Chlamydiae") GP.chl = prune_samples(sample_sums(GP.chl)\>=20, GP.chl)

GP.chl.merged = merge_taxa(GP.chl, taxa_names(GP.chl)\[1:5\])

ExportCSVData_GP.chl \<- cbind(GP.chl\@otu_table\@.Data, GP.chl\@tax_table\@.Data)

ExportCSVData_GP.chl.merged \<- cbind(GP.chl.merged\@otu_table\@.Data, GP.chl.merged\@tax_table\@.Data)

Áµ±ÂêàÂâç

![GP.chl](Phyloseq_RawData_Bac_images/tax_glom_v1.png)

Áµ±ÂêàÂæå

24341(ASVs)„Å´Áµ±Âêà„Åï„Çå„ÄÅ„É™„Éº„ÉâÊï∞„ÅåÂêàË®à„Åï„Çå„Çã

![GP.chl.merge](Phyloseq_RawData_Bac_images/tax_glom_v2.png)
:::

# PreProcess Methods

## Prevalence filtering in Taxa:

### Phylum

```{r}
#| label: Plots_Prevalence_Filtering_Phylum
#| eval: true
#| Include: false


library("phyloseq")
library("dplyr")
library("purrr")
library("ggplot2")
library("scales")
library("cowplot")

load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/phyloseq_Bacteria/Plot/Prevelence_inTaxa.RData")

prevdf_phylum_prevelence_plot
ggdraw(Phylum_legend_only)

```

### Class

```{r}
#| label: Plots_Prevalence_Filtering_Class
#| eval: true
#| Include: false


for (i in seq_along(plot_list_Class)) {
  if (is.null(plot_list_Class[[i]]) || is.null(plot_list_Class_legend[[i]])) next

  print(plot_list_Class[[i]])
  grid::grid.newpage()
  grid::grid.draw(plot_list_Class_legend[[i]])

}

```

### Order

```{r}
#| label: Plots_Prevalence_Filtering_Order
#| eval: true
#| Include: false


for (i in seq_along(plot_list_Order)) {
  if (is.null(plot_list_Order[[i]]) || is.null(plot_list_Order_legend[[i]])) next

  print(plot_list_Order[[i]])
  grid::grid.newpage()
  grid::grid.draw(plot_list_Order_legend[[i]])
}


```

### Family

```{r}
#| label: Plots_Prevalence_Filtering_Family
#| eval: true
#| Include: false

for (i in seq_along(plot_list_Family)) {
  if (is.null(plot_list_Family[[i]]) || is.null(plot_list_Family_legend[[i]])) next

  print(plot_list_Family[[i]])
  grid::grid.newpage()
  grid::grid.draw(plot_list_Family_legend[[i]])

}


```

### Genus

```{r}
#| label: Plots_Prevalence_Filtering_Genus
#| eval: true
#| Include: false

for (i in plot_list_Genus) {
    print(i)
}


```

```{r}
#| label: Prevalence filtering
#| eval: false
#| Include: false


# Prevalence filtering ----------------------

## Sample(Bac1~9)„Å´„Åä„ÅÑ„Å¶„ÄÅÂêÑASVs„ÅåÂá∫Áèæ„Åó„Åü(>0)SampleÊï∞(Â≠òÂú®È†ªÂ∫¶)„ÇíÂá∫Âäõ
## taxa_are_rows ‚Üí ASVs„ÅåË°å„ÅåÂàó„Åã„Å©„ÅÜ„Åã„ÇíÁ¢∫Ë™ç
prev0 = apply(X = otu_table(PhyseqData),
              MARGIN = ifelse(taxa_are_rows(PhyseqData), yes = 1, no = 2),
              FUN = function(x){sum(x > 0)})

## DataFrame„ÅÆ‰ΩúÊàê
prevdf = data.frame(Prevalence = prev0,
                    TotalAbundance = taxa_sums(PhyseqData),
                    tax_table(PhyseqData))
# write.csv(x = prevdf, file = "~/Documents/RStudio/Novogene/250503/export_csv/prevdf.csv")

## ÂÖ®ASVs„ÅÆ„Éô„ÇØ„Éà„É´(ID)„ÇíÂèñ„ÇäÂá∫„Åô ‚Üí TotalReads(ÂÖ®Sample„ÅÆASVs) > 10„Åß„Éï„Ç£„É´„Çø„É™„É≥„Ç∞
## keep_taxa <- taxa_names(PhyseqData)[taxa_sums(PhyseqData) >= 10]
## ps_filt <- prune_taxa(keep_taxa, PhyseqData)


## ÂêÑÂàÜÈ°û(Phylum)„É¨„Éô„É´„Åß„ÄÅtable„ÇíÂá∫Âäõ„ÄÅÂêÑÂàÜÈ°û(Phylum)„É¨„Éô„É´„ÅßÂàÜÈ°û„Åï„Çå„Åü„Éô„ÇØ„Éà„É´„Å´ÂØæ„Åó„Å¶„ÄÅ[ASVsÊï∞>5]„ÅßFiltering
## ASVs„ÅåÂÖ®Sample„ÅÆ‰∏≠„Åß„ÄÅ5Âõû‰ª•‰∏äÂá∫Áèæ„Åó„Å¶„ÅÑ„Å™„ÅÑ„Å®„ÅÑ„Åë„Å™„ÅÑ 
## ‰æã: ASV1 Bac1~4„ÅÆ„Åø„Åß>0„ÅÆÂ†¥Âêà„ÄÅÈô§Âéª„Åï„Çå„Çã
## ‚Üí ÊØîËºÉ„Åó„Åü„ÅÑÂá¶ÁêÜÂå∫ÂÜÖ„Åß„ÅÆSampleÊï∞„ÇíËÄÉÊÖÆ„Åô„Çã„Åì„Å®


## ÂÖ®sampleÊï∞„ÅÆ„ÅÜ„Å°„ÄÅ‰Ωïsample„ÅßÂêå‰∏Ä„ÅÆASVs„ÅåÂá∫Áèæ„Åó„Å¶„ÅÑ„Çå„Å∞„ÄÅËâØ„ÅÑ„Åã„ÅÆÂü∫Ê∫ñ„ÇíË®≠ÂÆö
## 0.1„ÅÆÂâ≤Âêà„Å†„Å®„ÄÅ9sample„ÅÆÂ†¥Âêà„ÄÅ0.9„Å®„Å™„Çä„ÄÅ1sample„ÅßÂá∫Áèæ„Åó„Å¶„ÅÑ„Çå„Å∞„ÅÑ„ÅÑ„Åì„Å®„Å´„Å™„Çã„Åü„ÇÅ„ÄÅ1„ÅßË®≠ÂÆö
prevalenceThreshold = 1
prevalenceThreshold


## Phylum ------------------------------------

keepPhyla = table(prevdf$Phylum)[(table(prevdf$Phylum) > 5)]
prevdf_phylum = subset(prevdf, Phylum %in% names(keepPhyla))
table(prevdf_phylum$Phylum)


prevdf_phylum_subset_prevelence_plot <- 
    ggplot(prevdf_phylum, aes(TotalAbundance, Prevalence, color = Phylum)) +
    geom_hline(yintercept = prevalenceThreshold, alpha = 0.5, linetype = 2) +
    geom_point(size = 1.5, alpha = 0.8) +
    scale_x_log10() +
    xlab("Total Abundance") + ylab("Prevalence") +
    facet_wrap(~Phylum) +
    theme(
        axis.title = element_text(size = 14, face = "bold", color = "black"),  
        axis.text = element_text(size = 12, face = "bold", color = "black"),   
        strip.text = element_text(size = 12, face = "bold", color = "black"),  
        panel.grid.minor = element_blank(),                                    
        legend.position = "none"                                               
    )


## Class -------------------------------------

keepClass = table(prevdf$Class)[(table(prevdf$Class) > 5)]
prevdf_Class = subset(prevdf, Class %in% names(keepClass))
table(prevdf_Class$Class)


## Âá∫ÁèæÊï∞È†Ü„Åßsort„Åó„Å¶‰∏ä‰Ωç40„Éª‰∏ã‰Ωç40„ÅßÂàÜÂâ≤
class_counts <- sort(table(prevdf_Class$Class), decreasing = TRUE)
top_classes <- names(class_counts)[1:41]
bottom_classes <- names(class_counts)[42:83]


## Top & Bottom„Åßsubset
prevdf_Class_top <- subset(prevdf_Class, Class %in% top_classes)
prevdf_Class_bottom <- subset(prevdf_Class, Class %in% bottom_classes)


prevdf_class_top40_prevelence_plot <- 
ggplot(prevdf_Class_top, aes(TotalAbundance, Prevalence, color = Class)) +
    geom_hline(yintercept = prevalenceThreshold, alpha = 0.5, linetype = 2) +
    geom_point(size = 1.5, alpha = 0.8) +
    scale_x_log10() +
    xlab("Total Abundance") + ylab("Prevalence") +
    facet_wrap(~Class, scales = "free_y") +
    theme(
        axis.title = element_text(size = 14, face = "bold", color = "black"),
        axis.text = element_text(size = 12, face = "bold", color = "black"),
        strip.text = element_text(size = 12, face = "bold", color = "black"),
        panel.grid.minor = element_blank(),
        legend.position = "none"
    )


prevdf_class_bottom40_prevelence_plot <- 
ggplot(prevdf_Class_bottom, aes(TotalAbundance, Prevalence, color = Class)) +
    geom_hline(yintercept = prevalenceThreshold, alpha = 0.5, linetype = 2) +
    geom_point(size = 1.5, alpha = 0.8) +
    scale_x_log10() + 
    xlab("Total Abundance") + ylab("Prevalence") +
    facet_wrap(~Class, scales = "free_y") +
    theme(
        axis.title = element_text(size = 14, face = "bold", color = "black"),
        axis.text = element_text(size = 12, face = "bold", color = "black"),
        strip.text = element_text(size = 12, face = "italic", color = "black"),
        panel.grid.minor = element_blank(),
        legend.position = "none"
    )

## Order -------------------------------------

keepOrder = table(prevdf$Order)[(table(prevdf$Order) > 5)]
prevdf_Order = subset(prevdf, Order %in% names(keepOrder))
table(prevdf_Order$Order)


class_counts_Order <- sort(table(na.omit(prevdf_Order$Order)), decreasing = TRUE)
length(class_counts_Order) 
order_chunks <- split(class_counts_Order, ceiling(seq_along(class_counts_Order) / 27)) # facet(~??)„ÅßÂàÜÂâ≤„Åó„Åü„ÅÑÊï∞„ÇíÊåáÂÆö


plot_list_order <- list() 
for (i in seq_along(order_chunks)) {
    subset_df <- subset(prevdf_Order, Order %in% names(order_chunks[[i]]))
    
    if (nrow(subset_df) == 0) next
    
    p <- ggplot(subset_df, aes(TotalAbundance, Prevalence, color = Order)) +
        geom_hline(yintercept = prevalenceThreshold, alpha = 0.7, linetype = 2, colour = "black") +
        geom_point(size = 1.5, alpha = 0.8) +
        scale_x_log10() +
        xlab("Total Abundance in Class") + ylab("Prevalence") +
        facet_wrap(~Order) +
        theme(axis.title = element_text(size = 18, face = "bold", color = "black"),
              axis.text = element_text(size = 12, face = "bold", color = "black"),
              strip.text = element_text(size = 14, face = "italic", color = "black"),
              panel.grid.minor = element_blank(),
              legend.position = "none")
    
    plot_list_order[[i]] <- p
}


# for (i in plot_list_order) {
#     print(i)
# }


## Family ------------------------------------

keepFamily = table(prevdf$Family)[(table(prevdf$Family) > 5)]
prevdf_Family = subset(prevdf, Family %in% names(keepFamily))
table(prevdf_Family$Family)

class_counts_Family <- sort(table(na.omit(prevdf_Family$Family)), decreasing = TRUE)
length(class_counts_Family) 
Family_chunks <- split(class_counts_Family, ceiling(seq_along(class_counts_Family) / 19)) # facet(~??)„ÅßÂàÜÂâ≤„Åó„Åü„ÅÑÊï∞„ÇíÊåáÂÆö


plot_list_Family <- list() 
for (i in seq_along(Family_chunks)) {
    subset_df <- subset(prevdf_Family, Family %in% names(Family_chunks[[i]]))
    
    if (nrow(subset_df) == 0) next
    
    p <- ggplot(subset_df, aes(TotalAbundance, Prevalence, color = Family)) +
        geom_hline(yintercept = prevalenceThreshold, alpha = 0.7, linetype = 2, colour = "black") +
        geom_point(size = 1.5, alpha = 0.8) +
        scale_x_log10() +
        xlab("Total Abundance in Family") + ylab("Prevalence") +
        facet_wrap(~Family) +
        theme(axis.title = element_text(size = 18, face = "bold", color = "black"),
              axis.text = element_text(size = 12, face = "bold", color = "black"),
              strip.text = element_text(size = 14, face = "italic", color = "black"),
              panel.grid.minor = element_blank(),
              legend.position = "none")
    
    plot_list_Family[[i]] <- p
}



## Genus -------------------------------------

keepGenus = table(prevdf$Genus)[(table(prevdf$Genus) > 5)] 
prevdf_Genus = subset(prevdf, Genus %in% names(keepGenus))
table(prevdf_Genus$Genus)

class_counts_Genus <- sort(table(na.omit(prevdf_Genus$Genus)), decreasing = TRUE)
length(class_counts_Genus) 
Genus_chunks <- split(class_counts_Genus, ceiling(seq_along(class_counts_Genus) / 11)) # facet(~??)„ÅßÂàÜÂâ≤„Åó„Åü„ÅÑÊï∞„ÇíÊåáÂÆö


plot_list_Genus <- list() 
for (i in seq_along(Genus_chunks)) {
    subset_df <- subset(prevdf_Genus, Genus %in% names(Genus_chunks[[i]]))
    
    if (nrow(subset_df) == 0) next
    
    p <- ggplot(subset_df, aes(TotalAbundance, Prevalence, color = Genus)) +
        geom_hline(yintercept = prevalenceThreshold, alpha = 0.7, linetype = 2, colour = "black") +
        geom_point(size = 1.5, alpha = 0.8) +
        scale_x_log10() +
        xlab("Total Abundance in Genus") + ylab("Prevalence") +
        facet_wrap(~Genus) +
        theme(axis.title = element_text(size = 18, face = "bold", color = "black"),
              axis.text = element_text(size = 12, face = "bold", color = "black"),
              strip.text = element_text(size = 14, face = "italic", color = "black"),
              panel.grid.minor = element_blank(),
              legend.position = "none")
    
    plot_list_Genus[[i]] <- p
}

rm(list = ls())

```

## Filtering Methods:

### ColorPallete:

```{r}
#| label: PhyloseqData & ColorPallete 
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center


load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/SaveObjects/PhyloseqData_Bac.RData")


PhyseqData
MetaData
track

### Color_Setting -----------------------------

library(phyloseq)
library(ggplot2)
library(cowplot)
library(dplyr)
library(scales)

colors <- c(
    "#1b9e77", "#d95f02", "#7570b3", "#e7298a", "#66a61e", 
    "#e6ab02", "#a6761d", "#666666", "#8dd3c7", "#ffffb3",
    "#bebada", "#fb8072", "#80b1d3", "#fdb462", "#b3de69", 
    "#fccde5", "#bc80bd", "#ccebc5", "#ffed6f", "#7fc97f", 
    "#fdc086", "#ffff99", "#386cb0", "#f0027f", "#bf5b17", 
    "#6a3d9a", "#cab2d6", "#ff7f00", "#b2df8a", "#a6cee3", 
    "#fb9a99", "#1f78b4", "#33a02c", "#b15928")

color_v2 <- c("#1F77B4", "#FF7F0E", "#2CA02C", "#D62728", "#9467BD", "#8C564B",
              "#E377C2", "#7F7F7F", "#BCBD22", "#17BECF", "#F0E442", "#9E14D2",
              "#7D8B33", "#B5F56D", "#3F51B5", "#D32F2F", "#0288D1", "#7B1FA2", 
              "#388E3C", "#FBC02D", "#0288D1", "#8BC34A", "#FF5722", "#8E24AA", 
              "#795548", "#9C27B0", "#3F51B5", "#4CAF50", "#FF9800", "#E91E63", 
              "#CDDC39")

colors_v3 <- c(colors, "#9999FF", "#FF99CC", "#66FFFF")

```

### Relative Abundunce:

#### Phylum Level

```{r}
#| label: Phylum Level
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center
#| 
load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/phyloseq_Bacteria/Output/PhyloseqData_Bacteria.RData")
### Phylum Level -------------------------------

PhyseqData_Phylum <- PhyseqData  |> 
    subset_taxa(Kingdom == "Bacteria") |> 
    tax_glom(taxrank = "Phylum") %>%                        # agglomerate at phylum level
    transform_sample_counts(function(x) {x/sum(x)} )  |>    # Transform to relative abundance
    psmelt()  |>                                            # Melt to long format
    filter(Abundance > 0.01)  |>                            # Filter out low(>1%) abundance taxa
    arrange(desc(Phylum))

unique(psmelt(PhyseqData)$Phylum)
length(unique(psmelt(PhyseqData)$Phylum))

Descending_Phylum <- PhyseqData_Phylum |> 
    group_by(Phylum)  |> 
    summarise(total_abundance = sum(Abundance, na.rm = TRUE))  |> 
    arrange(total_abundance)  |> 
    pull(Phylum)

PhyseqData_Phylum$Phylum <- factor(PhyseqData_Phylum$Phylum, levels = Descending_Phylum)

ggplot(PhyseqData_Phylum, aes(x = dps, y = Abundance, fill = Phylum)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors) +
    guides(fill = guide_legend(reverse = F, keywidth = 1, keyheight = 1.45)) +
    scale_y_continuous(labels = percent) +
    xlab("Days post Fungicide") +
    ylab("Relative Abundance (Phylum > 1%) \n") +
    theme(axis.title = element_text(size = 14, face = "bold", color = "black"),
          axis.text = element_text(size = 12, face = "bold", color = "black"),
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 10, color = "black"),
          legend.title = element_text(size = 14, face = "bold", color = "black", hjust = 0.5),
          legend.background = element_rect(fill = "gray90"),
          legend.key = element_rect(fill = "white", color = NA))

ggsave(filename = "Relative_abundance_Phylum.png", plot = last_plot(),
       width = 2800, height = 2520, dpi = 300, units = "px",
       path = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/png")


ggplot(PhyseqData_Phylum, aes(x = Sample.Name, y = Abundance, fill = Phylum)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors) +
    guides(fill = guide_legend(reverse = F, keywidth = 1.45, keyheight = 1.45)) +
    scale_y_continuous(labels = percent) +
    xlab("Samples") +
    ylab("Relative Abundance (Phylum > 1%) \n") +
    theme(axis.title = element_text(size = 16, face = "bold",color = "black"),
          axis.text = element_text(size = 10, color = "black"),
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 10, color = "black"),
          legend.title = element_text(size = 16, face = "bold", color = "black", hjust = 0.5),
          legend.background = element_rect(fill = "gray90"),
          legend.key = element_rect(fill = "white", color = NA),
          legend.key.size = unit(3, "cm"))

ggsave(filename = "Relative_abundance_Phylum_SampleName.png", plot = last_plot(),
       width = 2800, height = 2520, dpi = 300, units = "px",
       path = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/png")



```

#### Class Level

```{r}
#| label: Class Level
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center


PhyseqData_Class <- PhyseqData  |> 
    subset_taxa(Kingdom == "Bacteria") |> 
    tax_glom(taxrank = "Class")  |>                         # agglomerate at phylum level
    transform_sample_counts(function(x) {x/sum(x)} )  |>    # Transform to relative abundance
    psmelt()  |>                                            # Melt to long format
    filter(Abundance > 0.01)  |>                            # Filter out low(>1%) abundance taxa
    arrange(desc(Class))

unique(psmelt(PhyseqData)$Class)
length(unique(psmelt(PhyseqData)$Class))

Descending_Class <- PhyseqData_Class  |> 
    group_by(Class)  |> 
    summarise(total_abundance = sum(Abundance, na.rm = TRUE))  |> 
    arrange(total_abundance)  |> 
    pull(Class)

PhyseqData_Class$Class <- factor(PhyseqData_Class$Class, levels = Descending_Class)

ggplot(PhyseqData_Class, aes(x = dps, y = Abundance, fill = Class)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors) +
    guides(fill = guide_legend(reverse = F, keywidth = 1, keyheight = 1.45)) +
    scale_y_continuous(labels = percent) +
    xlab("Days post Fungicide") +
    ylab("Relative Abundance (Class > 1%) \n") +
    theme(axis.title = element_text(size = 14, face = "bold", color = "black"),
          axis.text = element_text(size = 12, face = "bold", color = "black"),
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 10, color = "black"),
          legend.title = element_text(size = 14, face = "bold", color = "black", hjust = 0.5),
          legend.background = element_rect(fill = "gray90"),
          legend.key = element_rect(fill = "white", color = NA))

ggsave(filename = "Relative_abundance_Class.png", plot = last_plot(),
       width = 2800, height = 2520, dpi = 300, units = "px",
       path = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/png")


ggplot(PhyseqData_Class, aes(x = Sample.Name, y = Abundance, fill = Class)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors) +
    guides(fill = guide_legend(reverse = F, keywidth = 1.45, keyheight = 1.45)) +
    scale_y_continuous(labels = percent) +
    xlab("Samples") +
    ylab("Relative Abundance (Class > 1%) \n") +
    theme(axis.title = element_text(size = 16, face = "bold",color = "black"),
          axis.text = element_text(size = 10, color = "black"),
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 10, color = "black"),
          legend.title = element_text(size = 16, face = "bold", color = "black", hjust = 0.5),
          legend.background = element_rect(fill = "gray90"),
          legend.key = element_rect(fill = "white", color = NA),
          legend.key.size = unit(3, "cm"))

ggsave(filename = "Relative_abundance_Class_SampleName.png", plot = last_plot(),
       width = 2800, height = 2520, dpi = 300, units = "px",
       path = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/png")




```

#### Order Level

```{r}
#| label: Order Level
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center


PhyseqData_Order <- PhyseqData  |> 
    subset_taxa(Kingdom == "Bacteria") |> 
    tax_glom(taxrank = "Order")  |>                         # agglomerate at phylum level
    transform_sample_counts(function(x) {x/sum(x)} )  |>    # Transform to relative abundance
    psmelt()  |>                                            # Melt to long format
    filter(Abundance > 0.01)  |>                            # Filter out low(>1%) abundance taxa
    arrange(desc(Order))

unique(psmelt(PhyseqData)$Order)
length(unique(psmelt(PhyseqData)$Order))

Descending_Order <- PhyseqData_Order  |> 
    group_by(Order)  |> 
    summarise(total_abundance = sum(Abundance, na.rm = TRUE))  |> 
    arrange(total_abundance)  |> 
    pull(Order)

PhyseqData_Order$Order <- factor(PhyseqData_Order$Order, levels = Descending_Order)

ggplot(PhyseqData_Order, aes(x = dps, y = Abundance, fill = Order)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors) +
    guides(fill = guide_legend(reverse = F, keywidth = 1, keyheight = 1.45)) +
    scale_y_continuous(labels = percent) +
    xlab("Days post Fungicide") +
    ylab("Relative Abundance (Order > 1%) \n") +
    theme(axis.title = element_text(size = 14, face = "bold", color = "black"),
          axis.text = element_text(size = 12, face = "bold", color = "black"),
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 10, color = "black"),
          legend.title = element_text(size = 14, face = "bold", color = "black", hjust = 0.5),
          legend.background = element_rect(fill = "gray90"),
          legend.key = element_rect(fill = "white", color = NA))

ggsave(filename = "Relative_abundance_Order.png", plot = last_plot(),
       width = 2800, height = 2520, dpi = 300, units = "px",
       path = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/png")


ggplot(PhyseqData_Order, aes(x = Sample.Name, y = Abundance, fill = Order)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors) +
    guides(fill = guide_legend(reverse = F, keywidth = 1.45, keyheight = 1.45)) +
    scale_y_continuous(labels = percent) +
    xlab("Samples") +
    ylab("Relative Abundance (Order > 1%) \n") +
    theme(axis.title = element_text(size = 16, face = "bold",color = "black"),
          axis.text = element_text(size = 10, color = "black"),
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 10, color = "black"),
          legend.title = element_text(size = 16, face = "bold", color = "black", hjust = 0.5),
          legend.background = element_rect(fill = "gray90"),
          legend.key = element_rect(fill = "white", color = NA),
          legend.key.size = unit(3, "cm"))

ggsave(filename = "Relative_abundance_Order_SampleName.png", plot = last_plot(),
       width = 2800, height = 2520, dpi = 300, units = "px",
       path = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/png")


```

#### Family Level

```{r}
#| label: Family Level
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center


PhyseqData_Family <- PhyseqData  |> 
    subset_taxa(Kingdom == "Bacteria") |> 
    tax_glom(taxrank = "Family")  |>                         # agglomerate at phylum level
    transform_sample_counts(function(x) {x/sum(x)} )  |>    # Transform to relative abundance
    psmelt()  |>                                            # Melt to long format
    filter(Abundance > 0.01)  |>                            # Filter out low(>1%) abundance taxa
    arrange(desc(Family))
    
unique(psmelt(PhyseqData)$Family)
length(unique(psmelt(PhyseqData)$Family))

Descending_Family <- PhyseqData_Family  |> 
    group_by(Family)  |> 
    summarise(total_abundance = sum(Abundance, na.rm = TRUE))  |> 
    arrange(total_abundance)  |> 
    pull(Family)

PhyseqData_Family$Family <- factor(PhyseqData_Family$Family, levels = Descending_Family)

ggplot(PhyseqData_Family, aes(x = dps, y = Abundance, fill = Family)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors) +
    guides(fill = guide_legend(reverse = F, keywidth = 1, keyheight = 1.45)) +
    scale_y_continuous(labels = percent) +
    xlab("Days post Fungicide") +
    ylab("Relative Abundance (Family > 1%) \n") +
    theme(axis.title = element_text(size = 14, face = "bold", color = "black"),
          axis.text = element_text(size = 12, face = "bold", color = "black"),
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 10, color = "black"),
          legend.title = element_text(size = 14, face = "bold", color = "black", hjust = 0.5),
          legend.background = element_rect(fill = "gray90"),
          legend.key = element_rect(fill = "white", color = NA))

ggsave(filename = "Relative_abundance_Family.png", plot = last_plot(),
       width = 2800, height = 2520, dpi = 300, units = "px",
       path = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/png")
       
       
ggplot(PhyseqData_Family, aes(x = Sample.Name, y = Abundance, fill = Family)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors) +
    guides(fill = guide_legend(reverse = F, keywidth = 1.45, keyheight = 1.45)) +
    scale_y_continuous(labels = percent) +
    xlab("Samples") +
    ylab("Relative Abundance (Family > 1%) \n") +
    theme(axis.title = element_text(size = 16, face = "bold",color = "black"),
          axis.text = element_text(size = 10, color = "black"),
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 10, color = "black"),
          legend.title = element_text(size = 16, face = "bold", color = "black", hjust = 0.5),
          legend.background = element_rect(fill = "gray90"),
          legend.key = element_rect(fill = "white", color = NA),
          legend.key.size = unit(3, "cm"))

ggsave(filename = "Relative_abundance_Family_SampleName.png", plot = last_plot(),
       width = 2800, height = 2520, dpi = 300, units = "px",
       path = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/png")


```

#### Genus Level

```{r}
#| label: Genus Level
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center

PhyseqData_Genus <- PhyseqData  |> 
    subset_taxa(Kingdom == "Bacteria") |> 
    tax_glom(taxrank = "Genus") |>                        
    transform_sample_counts(function(x) {x/sum(x)} )  |>   
    psmelt()  |>                                           
    filter(Abundance > 0.01)  |>                           
    arrange(desc(Genus))


unique(psmelt(PhyseqData)$Genus)
length(unique(psmelt(PhyseqData)$Genus))


Descending_Genus <- PhyseqData_Genus  |> 
    group_by(Genus)  |> 
    summarise(total_abundance = sum(Abundance, na.rm = TRUE))  |> 
    arrange(total_abundance)  |> 
    pull(Genus)

PhyseqData_Genus$Genus <- factor(PhyseqData_Genus$Genus, levels = Descending_Genus)

ggplot(PhyseqData_Genus, aes(x = dps, y = Abundance, fill = Genus)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors_v3) +
    guides(fill = guide_legend(reverse = F, keywidth = 0.7, keyheight = 1.45)) +
    scale_y_continuous(labels = percent) +
    xlab("Days post Fungicide") +
    ylab("Relative Abundance (Genus > 1%) \n") +
    theme(axis.title = element_text(size = 14, face = "bold", color = "black"),
          axis.text = element_text(size = 10, face = "bold", color = "black"),
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 10, color = "black"),
          legend.title = element_text(size = 14, face = "bold", color = "black", hjust = 0.5),
          legend.background = element_rect(fill = "gray90"),
          legend.key = element_rect(fill = "white", color = NA))

ggsave(filename = "Relative_abundance_Genus.png", plot = last_plot(),
       width = 2000, height = 1800, dpi = 300, units = "px",
       path = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/png")


ggplot(PhyseqData_Genus, aes(x = Sample.Name, y = Abundance, fill = Genus)) + 
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = colors_v3) +
    guides(fill = guide_legend(reverse = F, keywidth = 1.45, keyheight = 1.45)) +
    scale_y_continuous(labels = percent) +
    xlab("Samples") +
    ylab("Relative Abundance (Genus > 1%) \n") +
    theme(axis.title = element_text(size = 16, face = "bold",color = "black"),
          axis.text = element_text(size = 10, color = "black"),
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 10, color = "black"),
          legend.title = element_text(size = 16, face = "bold", color = "black", hjust = 0.5),
          legend.background = element_rect(fill = "gray90"),
          legend.key = element_rect(fill = "white", color = NA),
          legend.key.size = unit(3, "cm"))

ggsave(filename = "Relative_abundance_Genus_SampleName.png", plot = last_plot(),
       width = 2800, height = 2520, dpi = 300, units = "px",
       path = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/png")





```

::: callout-warning
## Abundance value transformation

[**It is usually necessary to transform microbiome count data to account for differences in library size, variance, scale, etc.**]{.mark-cyan}

`transform_sample_counts`:

The first argument to this function is the **phyloseq object** you want to transform, and the second argument is an R function that defines the transformation.

The R function is applied **sample-wise**, expecting that the first unnamed argument is a vector of taxa counts in the same order as the phyloseq object.

Additional arguments are passed on to the function specified in the second argument, providing an explicit means to include precomputed values, previously defined parameters/thresholds, or any other object that might be appropriate for computing the transformed values of interest.
:::

# Richness Index

```{r}
#| label: Richness Index
#| eval: false
#| Include: false

rm(list = ls())

library(phyloseq)
library(dplyr)
library(ggplot2)
library(ggpubr)

load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/phyloseq_Bacteria/Output/PhyloseqData_Bacteria.RData")


### Compare_RichnessIndex ---------------------
compare_means(Shannon ~ dps, data = alpha_df,
              method = "wilcox.test", label = "p.format")

compare_means(Shannon ~ dps, data = alpha_df,
              method = "t.test", label = "p.format")

compare_means(Shannon ~ dps, data = alpha_df,
              method = "anova", label = "p.format")

compare_means(Shannon ~ dps, data = alpha_df,
              method = "kruskal.test", label = "p.format")


### Plots_Richness ----------------------------
#### Compare_dps -------------------------------

# alpha_df |> 
#     ggboxplot(x = "dps", y = "Shannon", color = "dps",
#               palette = c("#00AFBB", "#E7B800", "#FC4E07"),
#               add = "jitter") + 
#     geom_pwc(method = "wilcox_test", label = "{p.format}{p.signif}") +
#     theme(axis.title = element_text(size = 14, face = "bold", color = "black"),
#           axis.text = element_text(size = 12, face = "bold", color = "black"),
#           panel.grid.minor = element_blank(),
#           legend.text = element_text(size = 10, color = "black"),
#           legend.title = element_text(size = 14, face = "bold", color = "black", hjust = 0.5))


## Â§öÊßòÊÄßÊåáÊï∞„ÅÆÁµ±Ë®àÂá¶ÁêÜ 
stat.test_dps <- compare_means(Shannon ~ dps, data = alpha_df, 
                               method = "wilcox.test", label = "p.format")

## y.position(pÂÄ§Ë°®Á§∫„ÅÆÈ´ò„Åï)
stat.test_dps$y.position <- seq(
    from = max(alpha_df$Shannon, na.rm = TRUE) * 1.05,
    by = max(alpha_df$Shannon, na.rm = TRUE) * 0.05,
    length.out = nrow(stat.test_dps))

## ggplots
ggplot(alpha_df, aes(x = dps, y = Shannon)) +
    geom_boxplot() +
    theme(
        legend.position = "top", 
        axis.title = element_text(size = 14, face = "bold", color = "black"),
        axis.text = element_text(size = 12, face = "bold", color = "black"),
        panel.background = element_rect(fill = "white"),
        panel.grid.major = element_line(color = "gray80"), 
        panel.grid.minor = element_line(color = "gray90")) + 
    geom_jitter(aes(color = Sample.Name), width = 0.08, size = 2.5, alpha = 0.6) +
    stat_pvalue_manual(stat.test_dps, label = "WilcoxTest {p.format} {p.signif}") +
    scale_color_manual(values = c("#E64B35",  
                                  "#4DBBD5",  
                                  "#00A087",  
                                  "#3C5488",  
                                  "#F39B7F",  
                                  "#8491B4",  
                                  "#91D1C2",  
                                  "#DC0000",  
                                  "#7E6148",  
                                  "#B09C85",  
                                  "#FFDC91",
                                  "#e7298a")) + 

ggsave(filename = "Richness_Shannon_dps.png", plot = last_plot(),
       width = 2800, height = 2520, dpi = 300, units = "px",
       path = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/png")

#### Compare_Fungicide.use ---------------------

alpha_df |> 
    ggboxplot(x = "Fungicide.use", y = "Shannon", color = "Fungicide.use",
              palette = c("#00AFBB", "#E7B800", "#FC4E07"),
              add = "jitter") + 
    geom_pwc(method = "wilcox_test", label = "{p.format}{p.signif}") +
    theme(axis.title = element_text(size = 14, face = "bold", color = "black"),
          axis.text = element_text(size = 12, face = "bold", color = "black"),
          panel.grid.minor = element_blank(),
          legend.text = element_text(size = 10, color = "black"),
          legend.title = element_text(size = 14, face = "bold", color = "black", hjust = 0.5))
          
    
stat.test <- compare_means(Shannon ~ Fungicide.use, data = alpha_df,
                           method = "wilcox.test", label = "p.format")

## y.position(pÂÄ§Ë°®Á§∫„ÅÆÈ´ò„Åï)
stat.test$y.position <- max(alpha_df$Shannon, na.rm = TRUE) * 1.05

## ggplots
ggplot(alpha_df, aes(x = Fungicide.use, y = Shannon)) +
    geom_boxplot() +
    geom_jitter(aes(color = Sample.Name), width = 0.1, size = 3, alpha = 0.6) +
    stat_pvalue_manual(stat.test, label = "„ÄêWilcox.Test„Äë {p.format}{p.signif}",
                       size = 5, bracket.size = 0.4) +
    scale_color_manual(values = c("#E64B35",  
                                  "#4DBBD5",  
                                  "#00A087",  
                                  "#3C5488",  
                                  "#F39B7F",  
                                  "#8491B4",  
                                  "#91D1C2",  
                                  "#DC0000",  
                                  "#7E6148",  
                                  "#B09C85",  
                                  "#FFDC91")) +
    theme(
        legend.position = "top", 
        axis.title = element_text(size = 18, face = "bold", color = "black"),
        axis.text = element_text(size = 14, face = "bold", color = "black"),
        panel.background = element_rect(fill = "white"),
        panel.grid.major = element_line(color = "gray80"), 
        panel.grid.minor = element_line(color = "gray90"))


ggsave(filename = "Richness_Shannon_Fungicide.use.png", plot = last_plot(),
       width = 2800, height = 2520, dpi = 300, units = "px",
       path = "~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/png")




```

# MicrobiotaProcessMethods:

```{r}
#| label: MicrobiotaProcessMethods
#| eval: true
#| Include: false
#| echo: true

# MPSE object Constructs
# mpdata <- PhyseqData |>
#     as.MPSE()


library("MicrobiotaProcess")
library("phyloseq")
library("dplyr")
library("purrr")
library("ggplot2")

load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/SaveObjects/mpdata.RData")

mpdata %<>% mp_rrarefy()

mpdata %<>% 
    mp_cal_rarecurve(
        .abundance = RareAbundance,
        chunks = 400
    )

```

## alpha diversity analysis

[**„É©„É≥„ÉÄ„É†„Çµ„É≥„Éó„É™„É≥„Ç∞„Åï„Çå„Åü„ÅÆ„Åå„ÄÅRareAbundance**]{.mark-red}

```{r}
#| label: alpha diversity analysis
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center



# mpdata <- PhyseqData |>
#     as.MPSE()

# save(mpdata,
#      file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/ËäùËçâ/NGS_consignment/Novogene/Data/250331/RData/SaveObjects/mpdata.RData")


# ‰ª•‰∏ã„ÅØÂêå„ÅòÊÑèÂë≥
## x <- x |> log()
## x %<>% log()

mpdata |> print(width=180)

p1 <- mpdata |> 
    mp_plot_rarecurve(
        .rare = RareAbundanceRarecurve,
        .alpha = Observe,
    )

# The Rarefaction of samples 
p1

# dev.off()


## calculate alpha index and visualization --------


mpdata %<>% 
    mp_cal_alpha(.abundance=RareAbundance) # Œ±Â§öÊßòÊÄßÊåáÊï∞„ÅÆÁÆóÂá∫
mpdata

f1 <- mpdata  |>  
    mp_plot_alpha(
        .group=dps, 
        .alpha=c(Observe, Chao1, ACE, Shannon, Simpson, Pielou)
    ) +
    scale_fill_manual(values=c("#00A087FF", "#3C5488FF","#cab2d6"), guide="none") +
    scale_color_manual(values=c("#00A087FF", "#3C5488FF","#cab2d6"), guide="none")

f2 <- mpdata |>
    mp_plot_alpha(
        .alpha=c(Observe, Chao1, ACE, Shannon, Simpson, Pielou)
    )

# The alpha diversity comparison
f1 / f2



# Èñ¢Êï∞„ÅÆÂÜÖÈÉ®Ë°®Á§∫
# MicrobiotaProcess::mp_plot_alpha
# showMethods(mp_plot_alpha)
# getMethod("mp_plot_alpha", signature = "MPSE")
# ggsignif::geom_signif()„Åß„ÄÅÊ§úÂÆöÊâãÊ≥ï„ÇíÂÆö„ÇÅ„Å¶„ÅÑ„Çã


```

## The visualization of taxonomy abundance

```{r}
#| label: The visualization of taxonomy abundance
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center


mpdata %<>%
    mp_cal_abundance( # for each samples
        .abundance = RareAbundance
    ) |>
    mp_cal_abundance( # for each groups 
        .abundance=RareAbundance,
        .group=dps
    )
mpdata

# „É©„É≥„ÉÄ„É†„Çµ„É≥„Éó„É™„É≥„Ç∞„Åï„Çå„ÅüRareAbundance„ÇíÁî®„ÅÑ„Å¶„ÅÑ„Çã„Åì„Å®„Å´Ê≥®ÊÑèÔºÅ
p1 <- mpdata |>
    mp_plot_abundance(
        .abundance=RareAbundance,
        .group=dps, 
        taxa.class = Phylum, 
        topn = 20,
        relative = TRUE
    )
# visualize the abundance (rarefied) of top 20 phyla for each sample
p2 <- mpdata |>
    mp_plot_abundance(
        .abundance=RareAbundance,
        .group=dps,
        taxa.class = Phylum,
        topn = 20,
        relative = FALSE
    )

# The relative abundunce and abundunce of phyla of all samples
p1
p2

# The relative abundance and abundance of phyla of all samples --------------------


h1 <- mpdata |>
    mp_plot_abundance(
        .abundance = RareAbundance,
        .group = dps,
        taxa.class = Phylum,
        relative = TRUE,
        topn = 20,
        geom = 'heatmap',
        features.dist = 'euclidean',
        features.hclust = 'average',
        sample.dist = 'bray',
        sample.hclust = 'average'
    ) 
h2 <- mpdata |>
    mp_plot_abundance(
        .abundance = RareAbundance,
        .group = dps,
        taxa.class = Phylum,
        relative = FALSE,
        topn = 20,
        geom = 'heatmap',
        features.dist = 'euclidean',
        features.hclust = 'average',
        sample.dist = 'bray',
        sample.hclust = 'average'
    )
# the character (scale or theme) of figure can be adjusted by set_scale_theme
# refer to the mp_plot_dist

# The relative abundunce and abundunce of phyla of all samples
aplot::plot_list(gglist=list(h1, h2), tag_levels="A")



# visualize the relative abundance of top 20 phyla for each .group (time) --------------------

p3 <- mpdata |>
    mp_plot_abundance(
        .abundance=RareAbundance, 
        .group=dps,
        taxa.class = Phylum,
        topn = 20,
        plot.group = TRUE
    )

# visualize the abundance of top 20 phyla for each .group (time)
p4 <- mpdata |>
    mp_plot_abundance(
        .abundance=RareAbundance,
        .group= dps,
        taxa.class = Phylum,
        topn = 20,
        relative = FALSE,
        plot.group = TRUE
    )

p3
p4


```

## Beta diversity analysis

```{r}
#| label: Beta diversity analysis
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center


# standardization
# mp_decostand wraps the decostand of vegan, which provides 
# many standardization methods for community ecology.
# default is hellinger, then the abundance processed will
# be stored to the assays slot. 
mpdata %<>% 
    mp_decostand(.abundance=Abundance)
mpdata

# calculate the distance between the samples.
# the distance will be generated a nested tibble and added to the
# colData slot.
mpdata %<>% mp_cal_dist(.abundance=hellinger, distmethod="bray")
mpdata

detach("package:phyloseq", unload=TRUE)


p5 <- mpdata |> mp_plot_dist(.distmethod = bray)
p5



# when .group is provided, the dot heatmap plot with group information will be return.
p6 <- mpdata |> MicrobiotaProcess::mp_plot_dist(.distmethod = bray, .group = dps)
# The scale or theme of dot heatmap plot can be adjusted using set_scale_theme function.
p6 |> set_scale_theme(
    x = scale_fill_manual(
        values=c("orange","deepskyblue","#1b9e77"), 
        guide = guide_legend(
            keywidth = 1, 
            keyheight = 0.5, 
            title.theme = element_text(size=8),
            label.theme = element_text(size=6))), 
    aes_var = dps) |> # specific the name of variable 
    set_scale_theme(
        x = scale_color_gradient(
            guide = guide_legend(keywidth = 0.5, keyheight = 0.5)),
        aes_var = bray) |> 
    set_scale_theme(
        x = scale_size_continuous(
            range = c(0.1, 3),
            guide = guide_legend(keywidth = 0.5, keyheight = 0.5)),
        aes_var = bray)


# when .group is provided and group.test is TRUE, the comparison of different groups will be returned
p7 <- mpdata |> mp_plot_dist(.distmethod = bray, .group = dps, group.test=TRUE, textsize=2)
p7


```

## The PCoA analysis

```{r}
#| label: The PCoA analysis 
#| eval: true
#| Include: false
#| out-width: 90%
#| fig-align: center


library(ggplot2)

mpdata %<>% 
    mp_cal_pcoa(.abundance = hellinger, distmethod="bray")

# The dimensions of ordination analysis will be added the colData slot (default).
mpdata

# We also can perform adonis or anosim to check whether it is significant to the dissimilarities of groups.
mpdata %<>%
    mp_adonis(.abundance = hellinger, .formula = ~dps, distmethod = "bray", permutations = 9999, action = "add")
mpdata %>% mp_extract_internal_attr(name = adonis)


p8 <- mpdata %>%
    mp_plot_ord(
        .ord = pcoa, 
        .group = dps, 
        .color = dps, 
        .size = 1.8,
        .alpha = 1,
        ellipse = TRUE,
        show.legend = FALSE # don't display the legend of stat_ellipse
    ) +
    scale_fill_manual(values=c("#00A087FF", "#3C5488FF", "orange")) +
    scale_color_manual(values=c("#00A087FF", "#3C5488FF", "orange")) 

# The size of point also can be mapped to other variables such as Observe, or Shannon 
# Then the alpha diversity and beta diversity will be displayed simultaneously.
p9 <- mpdata %>% 
    mp_plot_ord(
        .ord = pcoa, 
        .group = dps, 
        .color = dps, 
        .size = Shannon,
        ellipse = TRUE,
        show.legend = FALSE # don't display the legend of stat_ellipse 
    ) +
    scale_fill_manual(
        values = c("#00A087FF", "#3C5488FF","orange"), 
        guide = guide_legend(keywidth=0.6, keyheight=0.6, label.theme=element_text(size=6.5))
    ) +
    scale_color_manual(
        values=c("#00A087FF", "#3C5488FF", "orange"),
        guide = guide_legend(keywidth=0.6, keyheight=0.6, label.theme=element_text(size=6.5))
    ) +
    scale_size_continuous(
        range=c(0.5, 3),
        guide = guide_legend(keywidth=0.6, keyheight=0.6, label.theme=element_text(size=6.5))
    )


p8 + p9

```

## Hierarchical cluster analysis

```{r}
#| label: Hierarchical cluster analysis
#| eval: false
#| Include: false


mpdata %<>%
    mp_cal_clust(
        .abundance = hellinger, 
        distmethod = "bray",
        hclustmethod = "average", # (UPGAE)
        action = "add" # action is used to control which result will be returned
    )

mpdata



# if action = 'add', the result of hierarchical cluster will be added to the MPSE object
# mp_extract_internal_attr can extract it. It is a treedata object, so it can be visualized
# by ggtree.
sample.clust <- mpdata |>
    mp_extract_internal_attr(name='SampleClust')
sample.clust


library(ggtree)
library(ggtreeExtra)
library(ggplot2)


p10 <- ggtree(sample.clust) + 
    geom_tippoint(aes(color=dps)) +
    geom_tiplab(as_ylab = TRUE) +
    ggplot2::scale_x_continuous(expand=c(0, 0.01))


p10


phyla.tb <- mpdata |> 
    mp_extract_abundance(taxa.class = Phylum, topn = 30)

# The abundance of each samples is nested, it can be flatted using the unnest of tidyr.
phyla.tb %<>% tidyr::unnest(cols=RareAbundanceBySample) |> dplyr::rename(Phyla="label")
phyla.tb


p11 <- p10 + 
    geom_fruit(
        data = phyla.tb,
        geom = geom_col,
        mapping = aes(x = RelRareAbundanceBySample, 
                      y = Sample.Name,
                      fill = label
        ),
        orientation = "y",
        #offset = 0.4,
        pwidth = 3, 
        axis.params = list(axis = "x", 
                           title = "The relative abundance of phyla (%)",
                           title.size = 4,
                           text.size = 2, 
                           vjust = 1),
        grid.params = list()
    )


p11


```

## Biomarker discovery

### Transform Abundance value:

+------------------+------------------------------------------------------------------------------+-----------------------------------+
| ÂºïÊï∞             | ÊÑèÂë≥                                                                         | ‰æã                                |
+==================+==============================================================================+===================================+
| `.data`          | `MPSE`, `tbl_mpse`, `grouped_df_mpse`                                        | `mpdata`                          |
+------------------+------------------------------------------------------------------------------+-----------------------------------+
| `.abundance`     | Â≠òÂú®Èáè(colnames)\                                                            | `Abundance`                       |
+------------------+------------------------------------------------------------------------------+-----------------------------------+
| `min.abun`       | ÂêÑ„Çµ„É≥„Éó„É´„Å´„Åä„Åë„ÇãÊúÄÂ∞èÂ≠òÂú®Èáè                                                 | `1` ‚Üí ÂêÑ„Çµ„É≥„Éó„É´„Åß1‰ª•‰∏ä           |
+------------------+------------------------------------------------------------------------------+-----------------------------------+
| `min.prop`       | Taxa„ÅåÂ≠òÂú®„Åô„Åπ„Åç ÊúÄÂ∞è„Çµ„É≥„Éó„É´Ââ≤ÂêàÔºà0‚Äì1Ôºâ„Åæ„Åü„ÅØÊï∞ÂÄ§ÔºàÊï¥Êï∞Ôºâ‚Üí Precalence„ÅÆ„Åì„Å® | `0.1` ‚Üí ÂÖ®‰Ωì„ÅÆ10%„ÅÆ„Çµ„É≥„Éó„É´„Å´Â≠òÂú® |
+------------------+------------------------------------------------------------------------------+-----------------------------------+
| `include.lowest` | `TRUE` „ÅÆ„Å®„Åç `>= min.abun`„ÄÅ`FALSE` „Å™„Çâ `> min.abun`Ôºà„Éá„Éï„Ç©„É´„ÉàÔºâ         | `TRUE` „Å™„ÇâÈñæÂÄ§Âê´„ÇÄ               |
+------------------+------------------------------------------------------------------------------+-----------------------------------+

```{r}
#| label: Transform Abundance value
#| eval: false
#| Include: false

load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/SaveObjects/mpdata.RData")

mpdata %<>%
    mp_cal_abundance(.abundance = Abundance, action = "add",
                     relative = TRUE, force = TRUE)
mpdata %<>%
    mp_filter_taxa(.abundance = RelAbundanceBySample,
                   min.prop = 0.12,
                   min.abun = 0.1 # Áõ∏ÂØæÂ≠òÂú®Èáè 0.1% > Filtering  
    )

### Differentially abundant taxa --------------

mpdata %<>%
    mp_diff_analysis(
       .abundance = Abundance, # Â≠òÂú®Èáè„ÇíÊåáÂÆö (Áõ∏ÂØæÂ≠òÂú®Èáè or „É©„É≥„ÉÄ„É†„Çµ„É≥„Éó„É™„É≥„Ç∞Â≠òÂú®Èáè or Áµ∂ÂØæÂ≠òÂú®Èáè or Ê≠£Ë¶èÂåñÂ≠òÂú®Èáè)
       .group = dps, # ÊØîËºÉÂØæË±°
       first.test.alpha = 0.3, # ÊúâÊÑèÊÄß„ÅÆ‰∏ÄÊ¨°Âà§ÂÆö„Å´Áî®„ÅÑ„ÇãpÂÄ§„ÅÆ„Åó„Åç„ÅÑÂÄ§
       second.test.alpha = 0.05, action = "add", relative = TRUE # Â§öÈáçÊ§úÂÆöË£úÊ≠£Âæå„ÅÆ„Åó„Åç„ÅÑÂÄ§
    )

```

`mp_diff_analysis()`ÔºöÊØîËºÉ„Åó„Åü„ÅÑÂá¶ÁêÜÂå∫„Å´„ÅØ„ÄÅÊúÄ‰Ωé„Åß„ÇÇ 5 Sample„ÅØÂøÖË¶Å

# DESeq2

::: callout-note
## Reference

[Analyzing RNA-seq data with DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)\
[RNA-seq workflow: gene-level exploratory analysis and differential expression](https://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html)\
[DESeq2 with phyloseq](https://joey711.github.io/phyloseq-extensions/DESeq2.html)
:::

::: callout-important
## Analyzing RNA-seq data with DESeq2

It is important to provide count matrices as input for DESeq2‚Äôs statistical model (Love, Huber, and Anders 2014) to hold, as only the count values allow assessing the measurement precision correctly.

The DESeq2 model internally corrects for library size, so transformed or normalized values such as counts scaled by library size should not be used as input.

A DESeqDataSet object must have an associated design formula. The design formula expresses the variables which will be used in modeling.

The formula should be a tilde (\~) followed by the variables with plus signs between them (it will be coerced into an formula if it is not already).

The design can be changed later, however then all differential analysis steps should be repeated, as the design formula is used to estimate the dispersions and to estimate the log2 fold changes of the model.

## **Pre-filtering**

While it is not necessary to pre-filter low count genes before running the DESeq2 functions, there are two reasons which make pre-filtering useful:

by removing rows in which there are very few reads, we reduce the memory size of the dds data object, and we increase the speed of count modeling within DESeq2.

It can also improve visualizations, as features with no information for differential expression are not plotted in dispersion plots or MA-plots.

Here we perform pre-filtering to keep only rows that have a count of at **least 10** for a minimal number of samples
:::

::: callout-important
## Preparing quantification input to DESeq2

The values in the matrix should be counts or estimated counts of sequencing reads/fragments\
This is important for `DESeq2`‚Äôs statistical model to hold, as only counts allow assessing the measurement precision correctly\
It is important to never provide **counts that were pre-normalized for sequencing depth/library size, as the statistical model is most powerful when applied to un-normalized counts**, and is designed to account for library size differences internally\
The object class used by the DESeq2 package to store the read counts and the intermediate estimated quantities during statistical analysis is the **DESeqDataSet**, which will usually be represented in the code here as an object `dds`\
[We will now show 4 ways of constructing a **DESeqDataSet**, depending on what pipeline was used upstream of DESeq2 to generated counts or estimated counts:]{.mark-cyan}

-   From transcript abundance files and tximport
-   **From a count matrix**
-   From htseq-count files
-   From a SummarizedExperiment object

Alternatively, the function `DESeqDataSetFromMatrix` can be used if you already have a matrix of read counts prepared from another source
:::

## PhyseqObjectsDESeq2

```{r}
#| label: PhyseqObjectsDESeq2
#| eval: false
#| Include: false

rm(list = ls())
library("phyloseq")
library("DESeq2")

load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/SaveObjects/PhyloseqData.RData")

PhyseqData <- transform_sample_counts(PhyseqData, function(x) 100*(x / sum(x)))
# PhyseqData <- filter_taxa(PhyseqData, function(x) mean(x) > 0.1, TRUE)

# write.csv(t(PhyseqData@otu_table@.Data),
#            file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/ËäùËçâ/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/PhyseqData@otu_table@.Data.csv")


dps_dds = phyloseq_to_deseq2(PhyseqData, ~ dps)
dps_dds = DESeq(dps_dds, test="Wald", fitType="parametric")
res = results(dps_dds, cooksCutoff = FALSE)
alpha = 0.01
# write.csv(res, file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/ËäùËçâ/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/res.csv")
sigtab = res[which(res$padj < alpha), ] 
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(PhyseqData)[rownames(sigtab), ], "matrix"))
head(sigtab)



```

### Taxa Level in Family Level:

[**Ê≥®ÊÑè**]{.mark-red}

tax_glom(taxrank = "Family") ‚Üí ASVs„ÅÆÂàÜÈ°ûÁæ§„ÇíÁµ±Âêà„Åó„Å¶„ÅÑ„Çã„Åü„ÇÅ„ÄÅÁßë(Family)‰ª•‰∏ã„ÅÆÂàÜÈ°ûÁæ§„Åß„ÅÇ„ÇãÂ±û(Genus)„ÅåÁï∞„Å™„ÇãASVs„ÇÇ„ÄÅÊï∞Â≠ó„ÅåÂ∞è„Åï„ÅÑASVs„Å´Ê†ºÁ¥ç„Åï„Çå„Çã

‚Üí ASV 1 „ÅÆFamily„ÅåÈ°ïËëó„Å´Â≠òÂú®Èáè„ÅåÂ§âÂãï„Åó„Åü„Å®„Åó„Å¶„ÇÇ„ÄÅ„Åù„ÅÆ‰∏≠„ÅÆGenus„ÅåÂêåÊßò„ÅÆÂÇæÂêë„ÇíÁ§∫„Åó„Å¶„ÅÑ„Çã„Å®„ÅØÈôê„Çâ„Å™„ÅÑ!!!!

```{r}
#| label: Taxa Level in Family Level
#| eval: true
#| Include: false
#| message: false

### Taxa Level in Family Level --------------------------------
rm(list = ls())
load("~/Documents/RStudio/Novogene/250503/NGS_analysis_microbiome/RData/SaveObjects/PhyloseqData_Bac.RData")
library("phyloseq")
library("DESeq2")


#### Filtering ---------------------
PhyseqData_Family <- PhyseqData  |> 
    tax_glom(taxrank = "Family") |>                  # agglomerate at phylum level
    filter_taxa(function(x) mean(x) > 100, TRUE) |>  # ReadÊï∞„ÅåÁ¥Ñ100000„Åß„ÅÇ„Çä„ÄÅ0.1%„ÅÆReadsÊï∞ 
    phyloseq::subset_taxa(Kingdom == "Bacteria")               # Archaea„ÇíÈô§Âéª


# PhyseqData_Family.csv <- read.csv("~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/ËäùËçâ/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/Filtering_ASVs_Table/PhyseqData_Family.csv",
#             header = TRUE, sep = ",", fileEncoding = "UTF-8")

# write.csv(cbind(t(PhyseqData_Family@otu_table@.Data), PhyseqData_Family@tax_table@.Data),
#           file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/ËäùËçâ/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/Filtering_ASVs_Table/PhyseqData_Family.csv")




#### DESeq2 conversion & results table ----------------

dps_dds = phyloseq_to_deseq2(PhyseqData_Family, ~ dps)       # dpsÂàó(Êï£Â∏ÉÂæåÊó•Êï∞)„ÅßÊØîËºÉ
dps_dds = DESeq(dps_dds, test="Wald", fitType="parametric")



res = results(dps_dds, cooksCutoff = FALSE)
# write.csv(res, file = "~/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/ËäùËçâ/NGS_consignment/Novogene/Data/NGS_Analysis/physeqData_csv/Filtering_ASVs_Table/res.csv")
 


alpha = 0.01
sigtab = res[which(res$padj < alpha), ] 
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(PhyseqData)[rownames(sigtab), ], "matrix"))
head(sigtab)
dim(sigtab)

res

#### Results_ggplot ----------------------------
library("ggplot2")
theme_set(theme_bw())


# Phylum order
x = tapply(sigtab$log2FoldChange, sigtab$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtab$Phylum = factor(as.character(sigtab$Phylum), levels=names(x))
# Genus order
x = tapply(sigtab$log2FoldChange, sigtab$Genus, function(x) max(x))
x = sort(x, TRUE)
sigtab$Genus = factor(as.character(sigtab$Genus), levels=names(x))

# ggplot(sigtab, aes(x=Family, y=log2FoldChange, color=Phylum)) + geom_point(size=abs(sigtab$log2FoldChange)) + 
#   theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))



sigtab$Sign <- ifelse(sigtab$log2FoldChange < 0, "Negative", "Positive")
sigtab$ASV <- rownames(sigtab)
custom_colors_21 <- c(
    "#E41A1C", "#377EB8", "#4DAF4A", "#984EA3", "#FF7F00","#FFFF33", "#A65628", "#F781BF", "#999999", "#66C2A5",
    "#FC8D62", "#8DA0CB", "#E78AC3", "#A6D854", "#FFD92F","#E5C494", "#B3B3B3", "#1B9E77", "#D95F02", "#7570B3","#E7298A"
)


ggplot(sigtab, aes(x = Family, y = abs(sigtab$log2FoldChange), color = Phylum, shape = Sign)) +
    geom_point(size = 5) +
    geom_text(aes(label = ASV), vjust = -1, size = 3) + 
    scale_color_manual(values = custom_colors_21) +
    theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.5)) +
    scale_shape_manual(values = c("Positive" = 16, "Negative" = 17)) +  # ‰∏∏„Å®‰∏âËßí„Å™„Å©
    ylab("log2FoldChange") +
    xlab("Family") +
    theme(legend.position = "right")



```

### Taxa Level in Order Level:

```{r}
#| label: Taxa Level in Order Level
#| eval: false
#| Include: false



```

### Taxa Level in Class Level:

```{r}
#| label: Taxa Level in Class Level
#| eval: false
#| Include: false



```

### Taxa Level in Phylum Level:

```{r}
#| label: Taxa Level in Phylum Level
#| eval: false
#| Include: false



```

# SessionInfo

```{r}
#| label: SessionInfo
#| eval: true
#| Include: false


sessionInfo()


```
