---
title: "NGS解析_v1"
author: "saito ikuto"
date: "2025/03/31"
date-format: iso
format: 
  html:
    grid: 
      body-width: 1024px
    code-fold: true
    toc: true
    toc-expand: false
    toc-location: right
    toc-depth: 5
    number-sections: true
    embed-resources: false
    highlight-style: atom-one
    theme: yeti
    css: style.css
  pdf: default
---

# Chunk Options & SetDirectory

```{r}
#| label: ChunkOption
#| include: false
#| 

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Documents/RStudio/Quarto/Hyperlocal Variation in Soil Iron and the Rhizosphere Bacterial Community Determines Dollar Spot Development in Amenity Turfgrass/Rmd")

```

::: callout-tip
## 【Quarto HP】

[Shortcuts & Options](https://quarto.org/docs/visual-editor/options.html)

[Guide](https://quarto.org/docs/guide/)

⌘/ → Insert Anything

#\| eval: false → 毎回のRenderingでコードチャンクを評価しても、意味ないので、FALSE設定、状況に応じて、FALSEにすること
:::

# PackageList & SessionInfo

```{r}
#| message: false 
#| label: PackageList & SessionInfo


lib <- c("dada2", "ggplot2", "dplyr", "Biostrings", "phyloseq", "ShortRead")
for (i in lib) {
  print(i)
  print(packageVersion(i))
  library(i, quietly = TRUE, verbose = FALSE, warn.conflicts = FALSE, character.only = TRUE)
}

sessionInfo()

```

::: callout-note
# **Library Construction, Quality Control and Sequencing**

After the DNA sample passes the test, use the amplification primer to amplify, and purify the DNA through gel electrophoresis.

The obtained fragments were end repaired, A-tailed and further ligated with Illumina adapter. The fragments with adapters were PCR amplified, size selected, and purified

Novegene mainly uses two methods to check the library:

（1）AATI was used to check the integrity and size of the DNA fragments in the library.

（2）Effective concentration of the library was detected by QPCR.

Quantified libraries will be pooled and sequenced on Illumina platforms, according to effective library concentration and data amount required
:::

::: callout-tip
# Sequencing Data Format

The original raw data from Illumina platform are transformed to Sequenced Reads, known as Raw Data or RAW Reads, by base calling. Raw data are recorded in a FASTQ file, which contains sequencing reads and corresponding sequencing quality.

Every read in FASTQ format is stored in four lines, as indicated below (Cock P.J.A. et al. 2010):

@HWI-ST1276:71:C1162ACXX:1:1101:1208:2458 1:N:0:CGATGT

NAAGAACACGTTCGGTCACCTCAGCACACTTGTGAATGTCATGGGATCCAT

-   

#55???BBBBB?BA\@DEEFFCFFHHFFCFFHHHHHHHFAE0ECFFD/AEHH

Line 1 begins with a '\@' character and is followed by the Illumina Sequence Identifiers and an optional description.

Line 2 is the raw sequence of the read.

Line 3 begins with a '+' character and is optionally followed by the same sequence identifiers and descriptions as in Line 1.

Line 4 encodes the quality values for the bases in Line 2 and contains the same number of characters as the bases in the read.

The quality of Solexa sequencing bases is expressed in ASCII values. According to the different quality schemes used in sequencing, the methods for calculating the decimal quality value are also different. The common calculation methods are as follows:

Quality value = ASCII values - 64 or Quality value = ASCII values - 33

The range of base quality values for Solexa sequencing is \[0,40\], that is, ASCII values are expressed as ‘\[B, h\]’ or ‘\[#, I\]’.

Correspondence between sequencing error rate and sequencing quality value. Specifically, if the sequencing error rate is expressed in E and the base quality value is expressed in Q, the relationship is as follows: Q= -10 log10 (E)

The quality of sequencing data will be mainly distributed above Q20, so as to ensure the normal operation of subsequent high-level analysis.

[**According to the characteristics of sequencing technology, the base quality at the end of the sequencing segment is generally lower than that at the front**]{.underline}

## Explanation of Sequence Data Processing

For paired-end (PE) sequencing, every sample should have 2 data flies (read1 file and read2 file). These 2 files have the same line number, you could use 'wc -l' command to check the line number in Linux environment. [**The line number divide by 4 is the number of reads.**]{.underline}

The data size is the space occupied by the data in the hard disk. It's related to the format of the disk and compression ratio. And it has no influence on the quantity of sequenced bases. Thus, the size of read1 file may be unequal to the size of read2 file.
:::

::: callout-note
## Filtering and analysis methods

The original sequencing data contains adapter information, low-quality bases, and undetected bases (N bases), which will cause great interference to the subsequent bioinformatic analysis.

These interference information are removed by fine filtering methods, and the final data called clean data or clean reads. The data format of these files are exactly the same as Raw data.

[**Filtering and analysis methods**]{.underline}:

a.  Filting out the adapter sequence and barcode and primer sequence in reads.

b.  Use Flash(Mago, T., et al, 2011) software to combine the read pairs with overlap as Tags.

c.  Fastp(Bokulich NAetal.,2012) software is used to filter the combined data, and the sequences containing more N or more low-quality bases are filtered out.

d.  Filter out the chimeric sequences in the combined sequence(Edgar RC etal.,2011)

Combine Tags represents tags merged from rawPE reads. It merged two pair-end reads into a single-end tag of amplified region length through flash software.

Clean Tags represents tags after filtering. It carries out quality control according to fastp software to remove sequences with low quality,high error rate and excessive N bases.

Effective Tags represents tags after filtering chimera and can be finally used for subsequent analysis
:::

::: callout-warning
## Sequencing error rate will increase with the increase of sequencing length

What is the acceptable range of error rate? The sequencing of Novogene will strictly control the data quality

In general, the sequencing error rate of a single base position should be less than 1%, and the maximum should not exceed 6%.
:::

::: callout-tip
## fastq Check In Terminal

cd ./\~fastq/gz → fastqがあるディレクトリへ移動

`gunzip -c ??.fastq.gz | less`

@250331

3.1 \*.extendedFrags.fastq.gz【 reads after merging】

3.2 1.fastq.gz 【reads1 with barcode and primer removing】

3.3 2.fastq.gz 【reads2 with barcode and primer removing】

3.4 \* raw 1.fastq.gz 【reads1 with barcode and primer】

3.5 raw 2.fastq.gz 【reads2 with barcode and primer】

[**Terminal(CLI)上で、先に、\* raw 1.fastq.gzを見た後に、1.fastq.gzを確認し、barcode and primerを検索すると引っかかるため、注意が必要**]{.underline}

### Results directory:

Removed barcode and primer sequence (rawPE): result/01.RawData/Sample_Name/\*\_1(2).fastq.gz

**Combined sequence**: result/01.RawData/Sample_Name/\*.extendedFrags.fastq.gz

Barcodes and primers information: result/01.RawData/SampleSeq_info.xls

**Removed chimera sequence (Nochime)**: result/01.CleanData/Sample_Name/*.effective.fastq.gz,*.fna.gz

Customer provided data is not displayed in the table by default
:::

# **C**utadapt & FastQC Scripts

```{r}
#| label: CutadaptScripts
#| eval: false

# → barcode and primerは、毎NGSで異なるため、結果から参照(SampleSeq_info.xls)
# → ??.fastq.gzは、あらかじめbarcode and primerが除去されているため、そのまま下流に使用する

seqkit head -n 10 ???.fastq.gz


# Fug8 
# BarcodeSequence TACGGAACG / CCAGTTACATGCGT
# LinkerPrimerSequence GCATCGATGAAGAACGCAGC / TCCTCCGCTTATTGATATGC

# Bac8
# V4 300bp 515F: GTGCCAGCMGCCGCGGTAA 806R: GGACTACHVGGGTWTCTAAT
# BarcodeSequence AGCCAAGTAT / TGACCATCGACAT
# LinkerPrimerSequence GTGCCAGCMGCCGCGGTAA / GGACTACHVGGGTWTCTAAT

# FastQC CUI 
# /Users/ikutosaito/Library/CloudStorage/GoogleDrive-saito2022@patholab-meiji.jp/My Drive/芝草/NGS_consignment/Novogene/Data/250331/RawData_Bacteria
fastqc --outdir ../FastQC_Reports_RawData_Bacteria *.fastq.gz
fastqc --outdir ../FastQC_Reports_RawData_Fungi *.fastq.gz
fastqc --outdir ../FastQC_Reports_FilteringData_Bacteria *.fastq.gz
fastqc --outdir ../FastQC_Reports_FilteringData_Fungi *.fastq.gz
# → 先にoutput用のフォルダーを作成しておく

```

:::: callout-tip
## Conda & FastQC

M1 Macbook Air で環境構築しているため、注意

::: callout-tip
\# 仮想環境起動 & FastQC CUI

conda activate qiime2-amplicon-2023.9
:::
::::
